{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO, Entrez\n",
    "Entrez.email = 'adam.hockenberry@utexas.edu'\n",
    "import time\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This database might be great to use, actually. But it only became available in this easy to use form after NCBI re-did a few things and this project was largely complete.\n",
    "\n",
    "Worth considering (one day) diving deeper into what is available here. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First make the starting datatable better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19336, 20)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>SRA_Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Publications</th>\n",
       "      <th>Geo_Location</th>\n",
       "      <th>USA</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>GenBank_Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_050152</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-13T00:00:00Z</td>\n",
       "      <td>Escherichia virus P1</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>101660</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lobocka,M.B., Rusin,M., Bugajska,O., Dobruk,A....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Enterobacteria phage P7, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_050153</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-13T00:00:00Z</td>\n",
       "      <td>Escherichia phage vB_EcoM-Ro157c2YLVW</td>\n",
       "      <td>Punavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>80732</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Liao,Y.-T., Quintela,I.A., Bridges,D.F., Liu,F...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017-05-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia phage vB_EcoM-Ro157c2YLVW, complet...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_050154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-13T00:00:00Z</td>\n",
       "      <td>Escherichia phage D6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>91159</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gilcrease,E.B., Casjens,S.R.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia coli</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Escherichia phage D6, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_050143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage datas</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>60746</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Olsen,N.S., Forero-Junco,L., Kot,W., Hansen,L.H.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Denmark</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2017</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas phage datas, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_050144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage Epa14</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>65797</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Jacobs,A.C., Freyberger,H.R., Farlow,J., Watte...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas sp.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pseudomonas phage Epa14, complete genome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession SRA_Accession          Release_Date  \\\n",
       "0  NC_050152           NaN  2020-08-13T00:00:00Z   \n",
       "1  NC_050153           NaN  2020-08-13T00:00:00Z   \n",
       "2  NC_050154           NaN  2020-08-13T00:00:00Z   \n",
       "3  NC_050143           NaN  2020-08-10T00:00:00Z   \n",
       "4  NC_050144           NaN  2020-08-10T00:00:00Z   \n",
       "\n",
       "                                 Species       Genus      Family  Length  \\\n",
       "0                   Escherichia virus P1   Punavirus  Myoviridae  101660   \n",
       "1  Escherichia phage vB_EcoM-Ro157c2YLVW   Punavirus  Myoviridae   80732   \n",
       "2                   Escherichia phage D6         NaN  Myoviridae   91159   \n",
       "3                Pseudomonas phage datas  Pbunavirus  Myoviridae   60746   \n",
       "4                Pseudomonas phage Epa14  Pbunavirus  Myoviridae   65797   \n",
       "\n",
       "  Sequence_Type Nuc_Completeness Genotype Segment  \\\n",
       "0        RefSeq         complete      NaN     NaN   \n",
       "1        RefSeq         complete      NaN     NaN   \n",
       "2        RefSeq         complete      NaN     NaN   \n",
       "3        RefSeq         complete      NaN     NaN   \n",
       "4        RefSeq         complete      NaN     NaN   \n",
       "\n",
       "                                             Authors Publications  \\\n",
       "0  Lobocka,M.B., Rusin,M., Bugajska,O., Dobruk,A....          NaN   \n",
       "1  Liao,Y.-T., Quintela,I.A., Bridges,D.F., Liu,F...          NaN   \n",
       "2                       Gilcrease,E.B., Casjens,S.R.          NaN   \n",
       "3   Olsen,N.S., Forero-Junco,L., Kot,W., Hansen,L.H.          NaN   \n",
       "4  Jacobs,A.C., Freyberger,H.R., Farlow,J., Watte...          NaN   \n",
       "\n",
       "  Geo_Location  USA                         Host Isolation_Source  \\\n",
       "0          NaN  NaN           Enterobacteriaceae              NaN   \n",
       "1          USA  NaN                  Escherichia              NaN   \n",
       "2          NaN  NaN             Escherichia coli              NaN   \n",
       "3      Denmark  NaN  Pseudomonas aeruginosa PAO1              NaN   \n",
       "4          NaN  NaN              Pseudomonas sp.              NaN   \n",
       "\n",
       "  Collection_Date BioSample                                      GenBank_Title  \n",
       "0             NaN       NaN           Enterobacteria phage P7, complete genome  \n",
       "1      2017-05-03       NaN  Escherichia phage vB_EcoM-Ro157c2YLVW, complet...  \n",
       "2             NaN       NaN              Escherichia phage D6, complete genome  \n",
       "3            2017       NaN           Pseudomonas phage datas, complete genome  \n",
       "4             NaN       NaN           Pseudomonas phage Epa14, complete genome  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df = pd.read_csv('../../Data/Other_possible_dbs/NCBI_phage_db/all_complete_phage_info.csv')\n",
    "df = pd.read_csv('../../Data/Other_possible_dbs/NCBI_phage_db/all_complete_phage_info_11_2020.csv')\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Siphoviridae          7308\n",
       "Myoviridae            2972\n",
       "Microviridae          2839\n",
       "Autographiviridae     1148\n",
       "Podoviridae           1120\n",
       "Herelleviridae         394\n",
       "Drexlerviridae         321\n",
       "Demerecviridae         251\n",
       "Inoviridae             184\n",
       "Ackermannviridae       167\n",
       "Cystoviridae            46\n",
       "Leviviridae             39\n",
       "Chaseviridae            38\n",
       "Tectiviridae            26\n",
       "Corticoviridae           6\n",
       "Sphaerolipoviridae       4\n",
       "Plasmaviridae            2\n",
       "Name: Family, dtype: int64"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Family'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(14361, 20)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['Host'].isnull()==False]\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mycolicibacterium smegmatis MC2 155                       1788\n",
       "Escherichia coli                                           906\n",
       "Gordonia terrae                                            457\n",
       "Lactococcus lactis                                         401\n",
       "Klebsiella pneumoniae                                      368\n",
       "                                                          ... \n",
       "Aerococcus viridans                                          1\n",
       "Salmonella enterica subsp. enterica serovar Braenderup       1\n",
       "Pseudoalteromonas sp. YD6                                    1\n",
       "Salmonella enterica subsp. enterica serovar Saintpaul        1\n",
       "Schaalia odontolytica                                        1\n",
       "Name: Host, Length: 1019, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Better treatment of host taxonomy**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H43 str. T22\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O145:NM\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Nodularia\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Gordonia\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O104:H4\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli BL21(DE3)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Bacillus\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H7 str. EDL933\n",
      "Skipping\n",
      "[Bacillus] clarkii\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Yersinia enterocolitica (type O:3)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Streptomyces coelicolor A3(2)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Edwardsiella\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "[Brevibacterium] flavum\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Curvibacter sp. AEP1.3\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Streptococcus anginosus MAS624\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Morganella sp. (in: Bacteria)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Clostridium saccharoperbutylacetonicum N1-4(HMT)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Skipping\n",
      "Erwinia horticola\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H7\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O111:H2\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O103:H25 str. NIPH-11060424\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli DH5[alpha]\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Bacillus sp. (in: Bacteria)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Vibrio maritimus\n",
      "Vibrio campbellii ATCC BAA-1116\n",
      "Flavobacterium sp. LMO6\n",
      "Streptococcus\n",
      "Geobacillus stearothermophilus\n",
      "Streptococcus suis TL13\n",
      "Brucella abortus 544\n",
      "Corynebacterium glutamicum\n",
      "Staphylococcus hominis\n",
      "Streptococcus infantarius\n",
      "Aeromonas media\n",
      "Rheinheimera sp. BAL341\n",
      "Escherichia coli PA2\n",
      "Streptococcus henryi DSM 19005\n",
      "Cronobacter sakazakii\n",
      "Streptomyces azureus\n",
      "Streptococcus agalactiae ILRI005\n",
      "Burkholderia cepacia\n",
      "Streptococcus agalactiae GD201008-001\n",
      "Sinorhizobium meliloti 1021\n",
      "Pseudoalteromonas hodoensis\n",
      "Vibrio coralliilyticus\n",
      "Escherichia sp. MOD1-EC6098\n",
      "Ochrobactrum sp. POC9\n",
      "Streptococcus thermophilus ND03\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lysinibacillus sphaericus C3-41\n",
      "Thermus sp. TC4\n",
      "Lactococcus garvieae Lg2\n",
      "Pseudomonas syringae\n",
      "Morganella morganii\n",
      "Microcystis aeruginosa\n",
      "Enterococcus faecalis 62\n",
      "Synechococcus sp. WH 8018\n",
      "Fusobacterium nucleatum\n",
      "Streptococcus intermedius\n",
      "Vibrio cholerae O1\n",
      "Streptococcus anginosus\n",
      "Sphaerotilus natans\n",
      "Idiomarinaceae bacterium N2-2\n",
      "Vibrio metschnikovii\n",
      "Streptococcus pneumoniae BS458\n",
      "Acaryochloris marina\n",
      "Rhizobium gallicum\n",
      "Acinetobacter baumannii A118\n",
      "Streptomyces toxytricini\n",
      "Sodalis glossinidius\n",
      "Mycolicibacterium aichiense\n",
      "Escherichia coli C\n",
      "Streptococcus pneumoniae ST556\n",
      "Lactococcus lactis subsp. cremoris TIFN7\n",
      "Streptococcus pneumoniae GA47033\n",
      "Bacteroides intestinalis\n",
      "Polaribacter\n",
      "Streptomyces sp.\n",
      "Lactobacillus\n",
      "Epibacterium mobile\n",
      "Microbacterium natoriense\n",
      "Caulobacter vibrioides CB15\n",
      "Streptococcus equinus\n",
      "Trichormus variabilis\n",
      "Plesiomonas sp.\n",
      "Streptococcus parauberis KCTC 11980BP\n",
      "Streptococcus suis ST1\n",
      "Aeromonas rivipollensis\n",
      "Pseudomonas aeruginosa CHA\n",
      "Myxococcus xanthus\n",
      "Sphingomonas paucimobilis\n",
      "Escherichia coli H8\n",
      "Clostridioides difficile\n",
      "Parabacteroides merdae\n",
      "Streptococcus pneumoniae GA41538\n",
      "Thermobifida fusca\n",
      "Gordonia alkanivorans\n",
      "Prochlorococcus marinus\n",
      "Streptococcus gordonii\n",
      "Yersinia\n",
      "Marinomonas sp.\n",
      "Sphingobium sp. IP1\n",
      "Halomonas ventosae\n",
      "Mycobacteroides abscessus subsp. bolletii F1725\n",
      "Escherichia coli J53\n",
      "Streptococcus parauberis KRS-02083\n",
      "Nonlabens\n",
      "Streptococcus mitis SK564\n",
      "Tsukamurella paurometabola\n",
      "Shigella dysenteriae\n",
      "Streptococcus suis YB51\n",
      "Enterococcus\n",
      "Cellulophaga baltica NN016038\n",
      "Streptomyces virginiae\n",
      "Streptomyces avermitilis\n",
      "Dinoroseobacter shibae\n",
      "Campylobacter jejuni subsp. jejuni\n",
      "Streptococcus agalactiae BSU450\n",
      "Phormidium sp. MIS-PhA\n",
      "Parabacteroides distasonis\n",
      "Cronobacter\n",
      "Clostridium perfringens ATCC 13124\n",
      "Tenacibaculum geojense\n",
      "Staphylococcus aureus subsp. aureus N315\n",
      "Pseudomonas fluorescens group\n",
      "Burkholderia pseudomallei K96243\n",
      "Paracoccus marcusii\n",
      "Streptococcus suis 98HAH33\n",
      "Staphylococcus haemolyticus\n",
      "Lactococcus\n",
      "Pseudoalteromonas marina\n",
      "Brevundimonas subvibrioides ATCC 15264\n",
      "Erwinia pyrifoliae\n",
      "Alteromonas\n",
      "Salmonella enterica subsp. enterica serovar Typhimurium str. LT2\n",
      "Paracoccus sulfuroxidans\n",
      "Mesorhizobium loti\n",
      "Propionibacterium\n",
      "Erysipelothrix rhusiopathiae\n",
      "Aminobacter sp. MSH1\n",
      "Shewanella inventionis\n",
      "Rhodoferax sp. IMCC26218\n",
      "Rhodobacter\n",
      "Bacillus pumilus BL8\n",
      "Clostridium tetani\n",
      "Lactobacillus delbrueckii subsp. bulgaricus\n",
      "Salinibacter ruber\n",
      "Microbacterium foliorum\n",
      "Pseudomonas aeruginosa PA14\n",
      "Streptomyces tricolor\n",
      "Streptococcus anginosus F0211\n",
      "Synechococcus sp. WH 8102\n",
      "Stenotrophomonas sp.\n",
      "Haemophilus influenzae\n",
      "Streptococcus pyogenes NZ131\n",
      "Paracoccus\n",
      "Streptococcus pseudopneumoniae 5247\n",
      "Streptococcus pseudopneumoniae 22725\n",
      "Salmonella enterica subsp. enterica serovar Rissen\n",
      "Streptococcus uberis B190\n",
      "Vibrio sp.\n",
      "Salmonella enterica subsp. enterica serovar Saintpaul\n",
      "Streptococcus pneumoniae GA17328\n",
      "Streptococcus oralis ATCC 49296\n",
      "Oncorhynchus mykiss\n",
      "Shewanella fidelis\n",
      "Gordonia sp.\n",
      "Streptococcus agalactiae GB00640\n",
      "Lactobacillus plantarum\n",
      "Vibrio mediterranei\n",
      "Prochlorococcus marinus str. MIT 9313\n",
      "Pseudomonas chlororaphis\n",
      "Streptococcus hyovaginalis DSM 12219\n",
      "Synechococcus sp. WH 7803\n",
      "Streptococcus thermophilus 1F8CT\n",
      "Aeromonas salmonicida subsp. salmonicida\n",
      "Mycobacterium sp.\n",
      "Pectobacterium carotovorum\n",
      "Streptococcus parauberis KCTC 11537\n",
      "Photobacterium damselae\n",
      "Bacillus subtilis\n",
      "Bacillus thuringiensis serovar kurstaki\n",
      "Aggregatibacter actinomycetemcomitans D11S-1\n",
      "Bacillus amyloliquefaciens\n",
      "Acinetobacter johnsonii\n",
      "Candidatus Pelagibacter\n",
      "Streptococcus pneumoniae 23F\n",
      "Salinivibrio sp.\n",
      "Providencia sp.\n",
      "Streptococcus phocae subsp. salmonis\n",
      "Escherichia coli PA33\n",
      "Acholeplasma laidlawii\n",
      "Microbacterium nematophilum\n",
      "Achromobacter animicus\n",
      "Shigella\n",
      "Escherichia coli PA11\n"
     ]
    }
   ],
   "source": [
    "taxonomies_to_fetch = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "problematic_hosts = []\n",
    "for host in list(set(df['Host']))[:]:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) != 1:\n",
    "        problematic_hosts.append(host)\n",
    "        continue\n",
    "    record_id = record['IdList'][0]\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        problematic_hosts.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mycolicibacterium smegmatis    1978\n",
       "Escherichia coli               1187\n",
       "Salmonella enterica             423\n",
       "Pseudomonas aeruginosa          400\n",
       "Gordonia terrae                 382\n",
       "                               ... \n",
       "Salipiger profundus               1\n",
       "Pseudomonas sp. 1-1-1b            1\n",
       "Rothia dentocariosa               1\n",
       "Phaeobacter inhibens              1\n",
       "Microcystis elabens               1\n",
       "Name: Host_species_name, Length: 549, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_species_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11    12021\n",
       "1       270\n",
       "4        15\n",
       "Name: GeneticCode, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['GeneticCode'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fix the problematic hosts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Escherichia coli O157:H43 str. T22',\n",
       " 'Escherichia coli O145:NM',\n",
       " 'Nodularia',\n",
       " 'Gordonia',\n",
       " 'Escherichia coli O104:H4',\n",
       " 'Escherichia coli BL21(DE3)',\n",
       " 'Bacillus',\n",
       " 'Escherichia coli O157:H7 str. EDL933',\n",
       " '[Bacillus] clarkii',\n",
       " 'Yersinia enterocolitica (type O:3)',\n",
       " 'Streptomyces coelicolor A3(2)',\n",
       " 'Edwardsiella',\n",
       " '[Brevibacterium] flavum',\n",
       " 'Curvibacter sp. AEP1.3',\n",
       " 'Streptococcus anginosus MAS624',\n",
       " 'Morganella sp. (in: Bacteria)',\n",
       " 'Clostridium saccharoperbutylacetonicum N1-4(HMT)',\n",
       " 'Methylophilaceae bacterium IMCC19250',\n",
       " 'Erwinia horticola',\n",
       " 'Escherichia coli O157:H7',\n",
       " 'Escherichia coli O111:H2',\n",
       " 'Escherichia coli O103:H25 str. NIPH-11060424',\n",
       " 'Escherichia coli DH5[alpha]',\n",
       " 'Bacillus sp. (in: Bacteria)',\n",
       " 'Yersinia',\n",
       " 'Phormidium sp. MIS-PhA',\n",
       " 'Paracoccus']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(set(problematic_hosts)))\n",
    "problematic_hosts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Cases with special characters**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escherichia coli O157 H43 str. T22\n",
      "Escherichia coli O145 NM\n",
      "Nodularia\n",
      "Gordonia\n",
      "Escherichia coli O104 H4\n",
      "Escherichia coli BL21 DE3 \n",
      "Bacillus\n",
      "Escherichia coli O157 H7 str. EDL933\n",
      "Bacillus clarkii\n",
      "Yersinia enterocolitica  type O 3 \n",
      "Streptomyces coelicolor A3 2 \n",
      "Edwardsiella\n",
      "Brevibacterium flavum\n",
      "Curvibacter sp. AEP1.3\n",
      "Streptococcus anginosus MAS624\n",
      "Morganella sp.  in  Bacteria \n",
      "Clostridium saccharoperbutylacetonicum N1-4 HMT \n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Erwinia horticola\n",
      "Escherichia coli O157 H7\n",
      "Escherichia coli O111 H2\n",
      "Escherichia coli O103 H25 str. NIPH-11060424\n",
      "Escherichia coli DH5alpha\n",
      "Bacillus sp.  in  Bacteria \n",
      "Yersinia\n",
      "Phormidium sp. MIS-PhA\n",
      "Paracoccus\n"
     ]
    }
   ],
   "source": [
    "still_problematic = []\n",
    "for host in problematic_hosts:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    host = host.replace(':', ' ')\n",
    "    host = host.replace('[', '').replace(']', '')\n",
    "    host = host.replace('(', ' ').replace(')', ' ')\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) != 1:\n",
    "        still_problematic.append(host)\n",
    "        continue\n",
    "    record_id = record['IdList'][0]\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        still_problematic.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Nodularia',\n",
       " 'Gordonia',\n",
       " 'Bacillus',\n",
       " 'Edwardsiella',\n",
       " 'Curvibacter sp. AEP1.3',\n",
       " 'Streptococcus anginosus MAS624',\n",
       " 'Methylophilaceae bacterium IMCC19250',\n",
       " 'Erwinia horticola',\n",
       " 'Yersinia',\n",
       " 'Phormidium sp. MIS-PhA',\n",
       " 'Paracoccus']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(set(still_problematic)))\n",
    "still_problematic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Cases with annoying genuses**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodularia\n",
      "Gordonia\n",
      "Bacillus\n",
      "Edwardsiella\n",
      "Curvibacter sp. AEP1.3\n",
      "Streptococcus anginosus MAS624\n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Erwinia horticola\n",
      "Yersinia\n",
      "Phormidium sp. MIS-PhA\n",
      "Paracoccus\n"
     ]
    }
   ],
   "source": [
    "really_problematic = []\n",
    "for host in still_problematic:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    host = host.replace(':', ' ')\n",
    "    host = host.replace('[', '').replace(']', '')\n",
    "    host = host.replace('(', ' ').replace(')', ' ')\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) == 0:\n",
    "        really_problematic.append(host)\n",
    "        continue\n",
    "    temp_ids = record['IdList']\n",
    "    successful_ids = []\n",
    "    for record_id in temp_ids:\n",
    "        handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "        records = Entrez.read(handle)\n",
    "        if len(records) != 1:\n",
    "            really_problematic.append(host)\n",
    "            continue        \n",
    "        record = records[0]\n",
    "        lineage = record['LineageEx']\n",
    "        for i in lineage:\n",
    "            if i['Rank']=='superkingdom' and i['ScientificName'] =='Bacteria':\n",
    "                successful_ids.append(record_id)\n",
    "    if len(successful_ids) == 1:\n",
    "        record_id = successful_ids[0]\n",
    "    else:\n",
    "        really_problematic.append(host)\n",
    "    \n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        really_problematic.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "['Curvibacter sp. AEP1.3', 'Streptococcus anginosus MAS624', 'Methylophilaceae bacterium IMCC19250', 'Erwinia horticola', 'Phormidium sp. MIS-PhA']\n"
     ]
    }
   ],
   "source": [
    "print(len(really_problematic))\n",
    "print(really_problematic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "answers = [2685271, 1328, 2030816, 551, 1199]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n"
     ]
    }
   ],
   "source": [
    "taxonomies_to_fetch = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "problematic_hosts = []\n",
    "for host, answer in zip(really_problematic, answers):\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    record_id = str(answer)\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)    \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Write the new and improved file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bacteria     12211\n",
       "Eukaryota      270\n",
       "Archaea          2\n",
       "Name: Host_superkingdom_name, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_superkingdom_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12211, 33)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['Host_superkingdom_name']=='Bacteria']\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../Data/NCBI_phage_db/all_complete_phage_info_HOSTTAXONOMY_05152020.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the rows that I care about"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12211, 33)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045425</td>\n",
       "      <td>2019-12-16T00:00:00Z</td>\n",
       "      <td>Thermus virus OH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>5688</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>188787.0</td>\n",
       "      <td>Deinococci</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>Thermales</td>\n",
       "      <td>188786.0</td>\n",
       "      <td>Thermaceae</td>\n",
       "      <td>274.0</td>\n",
       "      <td>Thermus thermophilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_043767</td>\n",
       "      <td>2019-07-20T00:00:00Z</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "      <td>Rosebushvirus</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>67324</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>1760.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>85007.0</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>1762.0</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>1772.0</td>\n",
       "      <td>Mycolicibacterium smegmatis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_043027</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Bacillus virus PBS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>252197</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>91061.0</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>1385.0</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>186817.0</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_043028</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Xanthomonas virus Xf109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7190</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>347.0</td>\n",
       "      <td>Xanthomonas oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_043029</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Stenotrophomonas virus SMA6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7648</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>40324.0</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession          Release_Date                      Species  \\\n",
       "0  NC_045425  2019-12-16T00:00:00Z            Thermus virus OH3   \n",
       "1  NC_043767  2019-07-20T00:00:00Z    Mycobacterium virus TA17a   \n",
       "2  NC_043027  2019-06-28T00:00:00Z          Bacillus virus PBS1   \n",
       "3  NC_043028  2019-06-28T00:00:00Z      Xanthomonas virus Xf109   \n",
       "4  NC_043029  2019-06-28T00:00:00Z  Stenotrophomonas virus SMA6   \n",
       "\n",
       "           Genus        Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "0            NaN    Inoviridae    5688        RefSeq         complete   \n",
       "1  Rosebushvirus  Siphoviridae   67324        RefSeq         complete   \n",
       "2            NaN    Myoviridae  252197        RefSeq         complete   \n",
       "3            NaN    Inoviridae    7190        RefSeq         complete   \n",
       "4            NaN    Inoviridae    7648        RefSeq         complete   \n",
       "\n",
       "  Genotype Segment  ... Host_phylum_id     Host_phylum_name Host_class_id  \\\n",
       "0      NaN     NaN  ...         1297.0  Deinococcus-Thermus      188787.0   \n",
       "1      NaN     NaN  ...       201174.0       Actinobacteria        1760.0   \n",
       "2      NaN     NaN  ...         1239.0           Firmicutes       91061.0   \n",
       "3      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "4      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "\n",
       "       Host_class_name Host_order_id    Host_order_name Host_family_id  \\\n",
       "0           Deinococci       68933.0          Thermales       188786.0   \n",
       "1       Actinobacteria       85007.0  Corynebacteriales         1762.0   \n",
       "2              Bacilli        1385.0         Bacillales       186817.0   \n",
       "3  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "4  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "\n",
       "   Host_family_name  Host_species_id             Host_species_name  \n",
       "0        Thermaceae            274.0          Thermus thermophilus  \n",
       "1  Mycobacteriaceae           1772.0   Mycolicibacterium smegmatis  \n",
       "2       Bacillaceae           1423.0             Bacillus subtilis  \n",
       "3  Xanthomonadaceae            347.0            Xanthomonas oryzae  \n",
       "4  Xanthomonadaceae          40324.0  Stenotrophomonas maltophilia  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../../Data/Other_possible_dbs/NCBI_phage_db/all_complete_phage_info_HOSTTAXONOMY_05152020.tsv', sep='\\t')\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11150, 33)\n",
      "(11150, 33)\n",
      "(11141, 33)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_045425</td>\n",
       "      <td>2019-12-16T00:00:00Z</td>\n",
       "      <td>Thermus virus OH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>5688</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>188787.0</td>\n",
       "      <td>Deinococci</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>Thermales</td>\n",
       "      <td>188786.0</td>\n",
       "      <td>Thermaceae</td>\n",
       "      <td>274</td>\n",
       "      <td>Thermus thermophilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_043767</td>\n",
       "      <td>2019-07-20T00:00:00Z</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "      <td>Rosebushvirus</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>67324</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>1760.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>85007.0</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>1762.0</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>1772</td>\n",
       "      <td>Mycolicibacterium smegmatis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_043027</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Bacillus virus PBS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>252197</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>91061.0</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>1385.0</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>186817.0</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_043028</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Xanthomonas virus Xf109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7190</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>347</td>\n",
       "      <td>Xanthomonas oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_043029</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Stenotrophomonas virus SMA6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7648</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>40324</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession          Release_Date                      Species  \\\n",
       "0  NC_045425  2019-12-16T00:00:00Z            Thermus virus OH3   \n",
       "1  NC_043767  2019-07-20T00:00:00Z    Mycobacterium virus TA17a   \n",
       "2  NC_043027  2019-06-28T00:00:00Z          Bacillus virus PBS1   \n",
       "3  NC_043028  2019-06-28T00:00:00Z      Xanthomonas virus Xf109   \n",
       "4  NC_043029  2019-06-28T00:00:00Z  Stenotrophomonas virus SMA6   \n",
       "\n",
       "           Genus        Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "0            NaN    Inoviridae    5688        RefSeq         complete   \n",
       "1  Rosebushvirus  Siphoviridae   67324        RefSeq         complete   \n",
       "2            NaN    Myoviridae  252197        RefSeq         complete   \n",
       "3            NaN    Inoviridae    7190        RefSeq         complete   \n",
       "4            NaN    Inoviridae    7648        RefSeq         complete   \n",
       "\n",
       "  Genotype Segment  ... Host_phylum_id     Host_phylum_name Host_class_id  \\\n",
       "0      NaN     NaN  ...         1297.0  Deinococcus-Thermus      188787.0   \n",
       "1      NaN     NaN  ...       201174.0       Actinobacteria        1760.0   \n",
       "2      NaN     NaN  ...         1239.0           Firmicutes       91061.0   \n",
       "3      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "4      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "\n",
       "       Host_class_name Host_order_id    Host_order_name Host_family_id  \\\n",
       "0           Deinococci       68933.0          Thermales       188786.0   \n",
       "1       Actinobacteria       85007.0  Corynebacteriales         1762.0   \n",
       "2              Bacilli        1385.0         Bacillales       186817.0   \n",
       "3  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "4  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "\n",
       "   Host_family_name  Host_species_id             Host_species_name  \n",
       "0        Thermaceae              274          Thermus thermophilus  \n",
       "1  Mycobacteriaceae             1772   Mycolicibacterium smegmatis  \n",
       "2       Bacillaceae             1423             Bacillus subtilis  \n",
       "3  Xanthomonadaceae              347            Xanthomonas oryzae  \n",
       "4  Xanthomonadaceae            40324  Stenotrophomonas maltophilia  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = df[df['Host_species_id'].isnull()==False]\n",
    "print(df.shape)\n",
    "df['Host_species_id'] = df['Host_species_id'].astype(int)\n",
    "print(df.shape)\n",
    "df = df[df['GeneticCode']==11]\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1772       1978\n",
       "562        1279\n",
       "28901       423\n",
       "287         400\n",
       "2055        382\n",
       "1358        358\n",
       "573         251\n",
       "1771959     231\n",
       "1280        221\n",
       "104336      208\n",
       "1308        199\n",
       "1747        182\n",
       "1313        141\n",
       "470         116\n",
       "666         115\n",
       "1351        110\n",
       "1428        105\n",
       "1314        101\n",
       "1911         97\n",
       "599          79\n",
       "Name: Host_species_id, dtype: int64"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_species_id'].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1279, 33)\n"
     ]
    }
   ],
   "source": [
    "for host_taxid in [562]:\n",
    "    temp_df = df[df['Host_species_id']==host_taxid]\n",
    "    print(temp_df.shape)\n",
    "    assert len(set(list(temp_df['Accession']))) == len(list(temp_df['Accession']))\n",
    "    selected_accession_numbers = list(temp_df['Accession'])\n",
    "    all_genomes = SeqIO.parse('../../Data/Other_possible_dbs/NCBI_phage_db/all_complete_phage_nts.fasta', 'fasta')\n",
    "    for genome in all_genomes:\n",
    "        if genome.id in selected_accession_numbers:\n",
    "            with open('../../Data/Other_possible_dbs/NCBI_phage_db/phage_genomes/{}.fasta'.format(genome.id), 'w') as outfile:\n",
    "                SeqIO.write(genome, outfile, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Split apart the combined genome `fasta` file to write the genomes that I selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.read('../../Data/Other_possible_dbs/NCBI_phage_db/phage_genomes/NC_001416.fasta', 'fasta')\n",
    "# for genome_file in glob.glob('../Data/NCBI_phage_db/phage_genomes/*.fasta'):\n",
    "#     genome = SeqIO.read(genome_file, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73\n"
     ]
    }
   ],
   "source": [
    "hits = []\n",
    "all_cds = SeqIO.parse('../../Data/Other_possible_dbs/NCBI_phage_db/all_complete_phage_CDS.fasta', 'fasta')\n",
    "for cds in all_cds:\n",
    "    if genome.id in cds.id.split(':')[0]:\n",
    "        hits.append(cds)\n",
    "print(len(hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "73\n"
     ]
    }
   ],
   "source": [
    "print(len(hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "45192 4550\n",
      "False\n",
      "46186 4636\n",
      "False\n",
      "here\n",
      "29118 29285\n",
      "True\n",
      "34482 3503\n",
      "False\n",
      "here\n",
      "46459 46752\n",
      "True\n",
      "here\n",
      "47042 47575\n",
      "True\n",
      "47738 4794\n",
      "False\n",
      "here\n",
      "20147 20767\n",
      "True\n",
      "here\n",
      "22686 23918\n",
      "True\n",
      "here\n",
      "30839 31024\n",
      "True\n",
      "here\n",
      "31005 31196\n",
      "True\n",
      "here\n",
      "31169 31351\n",
      "True\n",
      "here\n",
      "34271 34357\n",
      "True\n",
      "here\n",
      "35037 35438\n",
      "True\n",
      "191 73\n",
      "False\n",
      "711 263\n",
      "False\n",
      "2633 283\n",
      "False\n",
      "2836 443\n",
      "False\n",
      "4418 573\n",
      "False\n",
      "5132 573\n",
      "False\n",
      "5747 607\n",
      "False\n",
      "6135 716\n",
      "False\n",
      "7202 760\n",
      "False\n",
      "7612 796\n",
      "False\n",
      "7977 855\n",
      "False\n",
      "8552 894\n",
      "False\n",
      "8955 969\n",
      "False\n",
      "9711 1013\n",
      "False\n",
      "10115 1054\n",
      "False\n",
      "10542 1310\n",
      "False\n",
      "13100 1342\n",
      "False\n",
      "13429 1412\n",
      "False\n",
      "14276 1487\n",
      "False\n",
      "14773 1544\n",
      "False\n",
      "15505 1890\n",
      "False\n",
      "18965 1958\n",
      "False\n",
      "19650 2085\n",
      "False\n",
      "21029 2197\n",
      "False\n",
      "21973 2255\n",
      "False\n",
      "here\n",
      "24509 25399\n",
      "True\n",
      "here\n",
      "25396 26973\n",
      "True\n",
      "here\n",
      "27812 28882\n",
      "True\n",
      "here\n",
      "28860 29078\n",
      "True\n",
      "here\n",
      "29374 29655\n",
      "True\n",
      "here\n",
      "29847 30395\n",
      "True\n",
      "here\n",
      "31348 32028\n",
      "True\n",
      "here\n",
      "32025 32810\n",
      "True\n",
      "here\n",
      "32816 33232\n",
      "True\n",
      "here\n",
      "33187 33330\n",
      "True\n",
      "here\n",
      "33299 33463\n",
      "True\n",
      "here\n",
      "33536 33904\n",
      "True\n",
      "here\n",
      "34087 34287\n",
      "True\n",
      "here\n",
      "35825 36259\n",
      "True\n",
      "here\n",
      "36275 37114\n",
      "True\n",
      "here\n",
      "37227 37940\n",
      "True\n",
      "38041 3824\n",
      "False\n",
      "38360 3865\n",
      "False\n",
      "38686 3958\n",
      "False\n",
      "39582 4028\n",
      "False\n",
      "40280 4057\n",
      "False\n",
      "40644 4108\n",
      "False\n",
      "41081 4195\n",
      "False\n",
      "41950 4212\n",
      "False\n",
      "42090 4227\n",
      "False\n",
      "42269 4243\n",
      "False\n",
      "42429 4304\n",
      "False\n",
      "43040 4324\n",
      "False\n",
      "43224 4388\n",
      "False\n",
      "43886 4450\n",
      "False\n",
      "44621 4481\n",
      "False\n",
      "45186 4550\n",
      "False\n",
      "45493 4596\n",
      "False\n",
      "45966 4642\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "for hit in hits:\n",
    "    reverse_complement = False\n",
    "    if 'join' in hit.id:\n",
    "        continue\n",
    "    if 'complement(' == hit.id[:11]:\n",
    "        print('here')\n",
    "        reverse_complement = True\n",
    "#         hit.id = hit.id[11:-1]\n",
    "    if len(hit.id.split(':')) == 2:\n",
    "        region = hit.id.split(':')[-1]\n",
    "        if len(region.split('..')) == 2:\n",
    "            start = int(region.split('..')[0])\n",
    "            stop = int(region.split('..')[-1][:-1])\n",
    "    print(start, stop)\n",
    "    print(genome.seq[start-1:stop]==hit.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make CDS datatables for each of those genomes by merging the genome `fasta` file with the CDS `fasta` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**et voila**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mycolicibacterium smegmatis    1978\n",
       "Escherichia coli               1279\n",
       "Salmonella enterica             423\n",
       "Pseudomonas aeruginosa          400\n",
       "Gordonia terrae                 382\n",
       "Lactococcus lactis              358\n",
       "Klebsiella pneumoniae           251\n",
       "Arthrobacter sp. ATCC 21022     231\n",
       "Staphylococcus aureus           221\n",
       "Microbacterium foliorum         208\n",
       "Streptococcus thermophilus      199\n",
       "Cutibacterium acnes             182\n",
       "Streptococcus pneumoniae        141\n",
       "Acinetobacter baumannii         116\n",
       "Vibrio cholerae                 115\n",
       "Enterococcus faecalis           110\n",
       "Bacillus thuringiensis          105\n",
       "Streptococcus pyogenes          101\n",
       "Streptomyces griseus             97\n",
       "Salmonella sp.                   79\n",
       "Name: Host_species_name, dtype: int64"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_species_name'].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11809</th>\n",
       "      <td>FJ483838</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3393</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11098</th>\n",
       "      <td>JX625144</td>\n",
       "      <td>2012-11-29T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage M</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3405</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1711</th>\n",
       "      <td>NC_019707</td>\n",
       "      <td>2012-12-07T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage M</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3405</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11808</th>\n",
       "      <td>FJ483837</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3412</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11810</th>\n",
       "      <td>FJ483839</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3458</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4272</th>\n",
       "      <td>MK817115</td>\n",
       "      <td>2019-06-29T00:00:00Z</td>\n",
       "      <td>Escherichia phage vB_EcoM_phAPEC6</td>\n",
       "      <td>Asteriusvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>352598</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6766</th>\n",
       "      <td>MH383160</td>\n",
       "      <td>2018-07-21T00:00:00Z</td>\n",
       "      <td>Escherichia phage UB</td>\n",
       "      <td>Asteriusvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>353081</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5544</th>\n",
       "      <td>MH791411</td>\n",
       "      <td>2019-02-06T00:00:00Z</td>\n",
       "      <td>Escherichia phage Ecwhy_1</td>\n",
       "      <td>Asteriusvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>354537</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428</th>\n",
       "      <td>MK327931</td>\n",
       "      <td>2019-03-25T00:00:00Z</td>\n",
       "      <td>Escherichia phage vB_EcoM_G17</td>\n",
       "      <td>Asteriusvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>370817</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6785</th>\n",
       "      <td>MH494197</td>\n",
       "      <td>2018-07-21T00:00:00Z</td>\n",
       "      <td>Escherichia phage CMSTMSU</td>\n",
       "      <td>Asteriusvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>386442</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1279 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession          Release_Date                            Species  \\\n",
       "11809   FJ483838  2009-09-09T00:00:00Z             Escherichia virus BZ13   \n",
       "11098   JX625144  2012-11-29T00:00:00Z             Enterobacteria phage M   \n",
       "1711   NC_019707  2012-12-07T00:00:00Z             Enterobacteria phage M   \n",
       "11808   FJ483837  2009-09-09T00:00:00Z             Escherichia virus BZ13   \n",
       "11810   FJ483839  2009-09-09T00:00:00Z             Escherichia virus BZ13   \n",
       "...          ...                   ...                                ...   \n",
       "4272    MK817115  2019-06-29T00:00:00Z  Escherichia phage vB_EcoM_phAPEC6   \n",
       "6766    MH383160  2018-07-21T00:00:00Z               Escherichia phage UB   \n",
       "5544    MH791411  2019-02-06T00:00:00Z          Escherichia phage Ecwhy_1   \n",
       "5428    MK327931  2019-03-25T00:00:00Z      Escherichia phage vB_EcoM_G17   \n",
       "6785    MH494197  2018-07-21T00:00:00Z          Escherichia phage CMSTMSU   \n",
       "\n",
       "               Genus       Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "11809      Levivirus  Leviviridae    3393       GenBank         complete   \n",
       "11098      Levivirus  Leviviridae    3405       GenBank         complete   \n",
       "1711       Levivirus  Leviviridae    3405        RefSeq         complete   \n",
       "11808      Levivirus  Leviviridae    3412       GenBank         complete   \n",
       "11810      Levivirus  Leviviridae    3458       GenBank         complete   \n",
       "...              ...          ...     ...           ...              ...   \n",
       "4272   Asteriusvirus   Myoviridae  352598       GenBank         complete   \n",
       "6766   Asteriusvirus   Myoviridae  353081       GenBank         complete   \n",
       "5544   Asteriusvirus   Myoviridae  354537       GenBank         complete   \n",
       "5428   Asteriusvirus   Myoviridae  370817       GenBank         complete   \n",
       "6785   Asteriusvirus   Myoviridae  386442       GenBank         complete   \n",
       "\n",
       "      Genotype Segment  ... Host_phylum_id Host_phylum_name Host_class_id  \\\n",
       "11809      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "11098      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "1711       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "11808      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "11810      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "...        ...     ...  ...            ...              ...           ...   \n",
       "4272       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "6766       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "5544       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "5428       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "6785       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "\n",
       "           Host_class_name Host_order_id   Host_order_name Host_family_id  \\\n",
       "11809  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "11098  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "1711   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "11808  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "11810  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "...                    ...           ...               ...            ...   \n",
       "4272   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "6766   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "5544   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "5428   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "6785   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "\n",
       "         Host_family_name  Host_species_id  Host_species_name  \n",
       "11809  Enterobacteriaceae              562   Escherichia coli  \n",
       "11098  Enterobacteriaceae              562   Escherichia coli  \n",
       "1711   Enterobacteriaceae              562   Escherichia coli  \n",
       "11808  Enterobacteriaceae              562   Escherichia coli  \n",
       "11810  Enterobacteriaceae              562   Escherichia coli  \n",
       "...                   ...              ...                ...  \n",
       "4272   Enterobacteriaceae              562   Escherichia coli  \n",
       "6766   Enterobacteriaceae              562   Escherichia coli  \n",
       "5544   Enterobacteriaceae              562   Escherichia coli  \n",
       "5428   Enterobacteriaceae              562   Escherichia coli  \n",
       "6785   Enterobacteriaceae              562   Escherichia coli  \n",
       "\n",
       "[1279 rows x 33 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df.sort_values('Length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>912</th>\n",
       "      <td>NC_028656</td>\n",
       "      <td>2015-12-17T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage VT2phi_272</td>\n",
       "      <td>Oslovirus</td>\n",
       "      <td>Podoviridae</td>\n",
       "      <td>65955</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1075</th>\n",
       "      <td>NC_021190</td>\n",
       "      <td>2015-05-28T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage phi80</td>\n",
       "      <td>Ravinvirus</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>46150</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1120</th>\n",
       "      <td>NC_023693</td>\n",
       "      <td>2014-11-25T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage phi92</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>148612</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2147</th>\n",
       "      <td>NC_010324</td>\n",
       "      <td>2008-02-11T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiEco32</td>\n",
       "      <td>Kuravirus</td>\n",
       "      <td>Podoviridae</td>\n",
       "      <td>77554</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2154</th>\n",
       "      <td>NC_010106</td>\n",
       "      <td>2007-12-04T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage phiEcoM-GJ1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>52975</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11987</th>\n",
       "      <td>EF380030</td>\n",
       "      <td>2007-02-19T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11988</th>\n",
       "      <td>EF380031</td>\n",
       "      <td>2007-02-19T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11989</th>\n",
       "      <td>EF380032</td>\n",
       "      <td>2007-02-19T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12022</th>\n",
       "      <td>DQ126339</td>\n",
       "      <td>2006-03-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiV10</td>\n",
       "      <td>Uetakevirus</td>\n",
       "      <td>Podoviridae</td>\n",
       "      <td>39104</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12138</th>\n",
       "      <td>AY135739</td>\n",
       "      <td>2003-08-02T00:00:00Z</td>\n",
       "      <td>Escherichia virus Wphi</td>\n",
       "      <td>Peduovirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>32684</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession          Release_Date                           Species  \\\n",
       "912    NC_028656  2015-12-17T00:00:00Z   Enterobacteria phage VT2phi_272   \n",
       "1075   NC_021190  2015-05-28T00:00:00Z        Enterobacteria phage phi80   \n",
       "1120   NC_023693  2014-11-25T00:00:00Z        Enterobacteria phage phi92   \n",
       "2147   NC_010324  2008-02-11T00:00:00Z        Escherichia virus phiEco32   \n",
       "2154   NC_010106  2007-12-04T00:00:00Z  Enterobacteria phage phiEcoM-GJ1   \n",
       "...          ...                   ...                               ...   \n",
       "11987   EF380030  2007-02-19T00:00:00Z         Escherichia virus phiX174   \n",
       "11988   EF380031  2007-02-19T00:00:00Z         Escherichia virus phiX174   \n",
       "11989   EF380032  2007-02-19T00:00:00Z         Escherichia virus phiX174   \n",
       "12022   DQ126339  2006-03-01T00:00:00Z          Escherichia virus phiV10   \n",
       "12138   AY135739  2003-08-02T00:00:00Z            Escherichia virus Wphi   \n",
       "\n",
       "                 Genus        Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "912          Oslovirus   Podoviridae   65955        RefSeq         complete   \n",
       "1075        Ravinvirus  Siphoviridae   46150        RefSeq         complete   \n",
       "1120               NaN    Myoviridae  148612        RefSeq         complete   \n",
       "2147         Kuravirus   Podoviridae   77554        RefSeq         complete   \n",
       "2154               NaN    Myoviridae   52975        RefSeq         complete   \n",
       "...                ...           ...     ...           ...              ...   \n",
       "11987  Sinsheimervirus  Microviridae    5386       GenBank         complete   \n",
       "11988  Sinsheimervirus  Microviridae    5386       GenBank         complete   \n",
       "11989  Sinsheimervirus  Microviridae    5386       GenBank         complete   \n",
       "12022      Uetakevirus   Podoviridae   39104       GenBank         complete   \n",
       "12138       Peduovirus    Myoviridae   32684       GenBank         complete   \n",
       "\n",
       "      Genotype Segment  ... Host_phylum_id Host_phylum_name Host_class_id  \\\n",
       "912        NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "1075       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "1120       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "2147       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "2154       NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "...        ...     ...  ...            ...              ...           ...   \n",
       "11987      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "11988      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "11989      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "12022      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "12138      NaN     NaN  ...         1224.0   Proteobacteria        1236.0   \n",
       "\n",
       "           Host_class_name Host_order_id   Host_order_name Host_family_id  \\\n",
       "912    Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "1075   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "1120   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "2147   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "2154   Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "...                    ...           ...               ...            ...   \n",
       "11987  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "11988  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "11989  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "12022  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "12138  Gammaproteobacteria       91347.0  Enterobacterales          543.0   \n",
       "\n",
       "         Host_family_name  Host_species_id  Host_species_name  \n",
       "912    Enterobacteriaceae              562   Escherichia coli  \n",
       "1075   Enterobacteriaceae              562   Escherichia coli  \n",
       "1120   Enterobacteriaceae              562   Escherichia coli  \n",
       "2147   Enterobacteriaceae              562   Escherichia coli  \n",
       "2154   Enterobacteriaceae              562   Escherichia coli  \n",
       "...                   ...              ...                ...  \n",
       "11987  Enterobacteriaceae              562   Escherichia coli  \n",
       "11988  Enterobacteriaceae              562   Escherichia coli  \n",
       "11989  Enterobacteriaceae              562   Escherichia coli  \n",
       "12022  Enterobacteriaceae              562   Escherichia coli  \n",
       "12138  Enterobacteriaceae              562   Escherichia coli  \n",
       "\n",
       "[87 rows x 33 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df[temp_df['Species'].str.contains('phi')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "88007.0"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.percentile(temp_df['Length'], 75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Accession', 'Release_Date', 'Species', 'Genus', 'Family', 'Length',\n",
       "       'Sequence_Type', 'Nuc_Completeness', 'Genotype', 'Segment', 'Authors',\n",
       "       'Publications', 'Geo_Location', 'Host', 'Isolation_Source',\n",
       "       'Collection_Date', 'BioSample', 'GenBank_Title', 'GeneticCode',\n",
       "       'Host_genus_id', 'Host_genus_name', 'Host_superkingdom_id',\n",
       "       'Host_superkingdom_name', 'Host_phylum_id', 'Host_phylum_name',\n",
       "       'Host_class_id', 'Host_class_name', 'Host_order_id', 'Host_order_name',\n",
       "       'Host_family_id', 'Host_family_name', 'Host_species_id',\n",
       "       'Host_species_name'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11    1279\n",
       "Name: GeneticCode, dtype: int64"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df['GeneticCode'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11    11141\n",
       "Name: GeneticCode, dtype: int64"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['GeneticCode'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Siphoviridae          6079\n",
       "Myoviridae            1951\n",
       "Podoviridae            912\n",
       "Microviridae           479\n",
       "Autographiviridae      460\n",
       "Herelleviridae         298\n",
       "Drexlerviridae         184\n",
       "Demerecviridae         112\n",
       "Ackermannviridae        95\n",
       "Inoviridae              81\n",
       "Cystoviridae            31\n",
       "Leviviridae             22\n",
       "Tectiviridae            12\n",
       "Sphaerolipoviridae       4\n",
       "Corticoviridae           2\n",
       "Plasmaviridae            1\n",
       "Name: Family, dtype: int64"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Family'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df = df[df['Host_species_id']==562]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11809</th>\n",
       "      <td>FJ483838</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3393</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11098</th>\n",
       "      <td>JX625144</td>\n",
       "      <td>2012-11-29T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage M</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3405</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1711</th>\n",
       "      <td>NC_019707</td>\n",
       "      <td>2012-12-07T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage M</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3405</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11808</th>\n",
       "      <td>FJ483837</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3412</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11810</th>\n",
       "      <td>FJ483839</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3458</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2512</th>\n",
       "      <td>NC_001426</td>\n",
       "      <td>1993-04-28T00:00:00Z</td>\n",
       "      <td>Escherichia virus BZ13</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3466</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11173</th>\n",
       "      <td>JQ966307</td>\n",
       "      <td>2012-09-18T00:00:00Z</td>\n",
       "      <td>Escherichia virus MS2</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3525</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5791</th>\n",
       "      <td>MK213795</td>\n",
       "      <td>2019-01-08T00:00:00Z</td>\n",
       "      <td>Escherichia virus MS2</td>\n",
       "      <td>Levivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>3526</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2486</th>\n",
       "      <td>NC_001890</td>\n",
       "      <td>1998-04-25T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4215</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11811</th>\n",
       "      <td>FJ483840</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4218</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11813</th>\n",
       "      <td>FJ483842</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4218</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11814</th>\n",
       "      <td>FJ483843</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4218</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11815</th>\n",
       "      <td>FJ483844</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4219</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11812</th>\n",
       "      <td>FJ483841</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus Qbeta</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4221</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11816</th>\n",
       "      <td>FJ539132</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus FI</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4241</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11817</th>\n",
       "      <td>FJ539133</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus FI</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4243</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11819</th>\n",
       "      <td>FJ539135</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus FI</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4273</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11818</th>\n",
       "      <td>FJ539134</td>\n",
       "      <td>2009-09-09T00:00:00Z</td>\n",
       "      <td>Escherichia virus FI</td>\n",
       "      <td>Allolevivirus</td>\n",
       "      <td>Leviviridae</td>\n",
       "      <td>4273</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4462</th>\n",
       "      <td>MK629526</td>\n",
       "      <td>2019-05-04T00:00:00Z</td>\n",
       "      <td>Escherichia phage Lilleen</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5342</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11577</th>\n",
       "      <td>HM753667</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11983</th>\n",
       "      <td>EF380026</td>\n",
       "      <td>2007-02-19T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11578</th>\n",
       "      <td>HM753668</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11586</th>\n",
       "      <td>HM753676</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11576</th>\n",
       "      <td>HM753666</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11575</th>\n",
       "      <td>HM753665</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11574</th>\n",
       "      <td>HM753664</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11569</th>\n",
       "      <td>HM753659</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11568</th>\n",
       "      <td>HM753658</td>\n",
       "      <td>2011-07-01T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11989</th>\n",
       "      <td>EF380032</td>\n",
       "      <td>2007-02-19T00:00:00Z</td>\n",
       "      <td>Escherichia virus phiX174</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10506</th>\n",
       "      <td>KJ997912</td>\n",
       "      <td>2014-08-06T00:00:00Z</td>\n",
       "      <td>Enterobacteria phage phiX174 sensu lato</td>\n",
       "      <td>Sinsheimervirus</td>\n",
       "      <td>Microviridae</td>\n",
       "      <td>5386</td>\n",
       "      <td>GenBank</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>562</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows  33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Accession          Release_Date  \\\n",
       "11809   FJ483838  2009-09-09T00:00:00Z   \n",
       "11098   JX625144  2012-11-29T00:00:00Z   \n",
       "1711   NC_019707  2012-12-07T00:00:00Z   \n",
       "11808   FJ483837  2009-09-09T00:00:00Z   \n",
       "11810   FJ483839  2009-09-09T00:00:00Z   \n",
       "2512   NC_001426  1993-04-28T00:00:00Z   \n",
       "11173   JQ966307  2012-09-18T00:00:00Z   \n",
       "5791    MK213795  2019-01-08T00:00:00Z   \n",
       "2486   NC_001890  1998-04-25T00:00:00Z   \n",
       "11811   FJ483840  2009-09-09T00:00:00Z   \n",
       "11813   FJ483842  2009-09-09T00:00:00Z   \n",
       "11814   FJ483843  2009-09-09T00:00:00Z   \n",
       "11815   FJ483844  2009-09-09T00:00:00Z   \n",
       "11812   FJ483841  2009-09-09T00:00:00Z   \n",
       "11816   FJ539132  2009-09-09T00:00:00Z   \n",
       "11817   FJ539133  2009-09-09T00:00:00Z   \n",
       "11819   FJ539135  2009-09-09T00:00:00Z   \n",
       "11818   FJ539134  2009-09-09T00:00:00Z   \n",
       "4462    MK629526  2019-05-04T00:00:00Z   \n",
       "11577   HM753667  2011-07-01T00:00:00Z   \n",
       "11983   EF380026  2007-02-19T00:00:00Z   \n",
       "11578   HM753668  2011-07-01T00:00:00Z   \n",
       "11586   HM753676  2011-07-01T00:00:00Z   \n",
       "11576   HM753666  2011-07-01T00:00:00Z   \n",
       "11575   HM753665  2011-07-01T00:00:00Z   \n",
       "11574   HM753664  2011-07-01T00:00:00Z   \n",
       "11569   HM753659  2011-07-01T00:00:00Z   \n",
       "11568   HM753658  2011-07-01T00:00:00Z   \n",
       "11989   EF380032  2007-02-19T00:00:00Z   \n",
       "10506   KJ997912  2014-08-06T00:00:00Z   \n",
       "\n",
       "                                       Species            Genus        Family  \\\n",
       "11809                   Escherichia virus BZ13        Levivirus   Leviviridae   \n",
       "11098                   Enterobacteria phage M        Levivirus   Leviviridae   \n",
       "1711                    Enterobacteria phage M        Levivirus   Leviviridae   \n",
       "11808                   Escherichia virus BZ13        Levivirus   Leviviridae   \n",
       "11810                   Escherichia virus BZ13        Levivirus   Leviviridae   \n",
       "2512                    Escherichia virus BZ13        Levivirus   Leviviridae   \n",
       "11173                    Escherichia virus MS2        Levivirus   Leviviridae   \n",
       "5791                     Escherichia virus MS2        Levivirus   Leviviridae   \n",
       "2486                   Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11811                  Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11813                  Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11814                  Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11815                  Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11812                  Escherichia virus Qbeta    Allolevivirus   Leviviridae   \n",
       "11816                     Escherichia virus FI    Allolevivirus   Leviviridae   \n",
       "11817                     Escherichia virus FI    Allolevivirus   Leviviridae   \n",
       "11819                     Escherichia virus FI    Allolevivirus   Leviviridae   \n",
       "11818                     Escherichia virus FI    Allolevivirus   Leviviridae   \n",
       "4462                 Escherichia phage Lilleen              NaN  Microviridae   \n",
       "11577                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11983                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11578                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11586                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11576                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11575                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11574                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11569                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11568                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "11989                Escherichia virus phiX174  Sinsheimervirus  Microviridae   \n",
       "10506  Enterobacteria phage phiX174 sensu lato  Sinsheimervirus  Microviridae   \n",
       "\n",
       "       Length Sequence_Type Nuc_Completeness Genotype Segment  ...  \\\n",
       "11809    3393       GenBank         complete      NaN     NaN  ...   \n",
       "11098    3405       GenBank         complete      NaN     NaN  ...   \n",
       "1711     3405        RefSeq         complete      NaN     NaN  ...   \n",
       "11808    3412       GenBank         complete      NaN     NaN  ...   \n",
       "11810    3458       GenBank         complete      NaN     NaN  ...   \n",
       "2512     3466        RefSeq         complete      NaN     NaN  ...   \n",
       "11173    3525       GenBank         complete      NaN     NaN  ...   \n",
       "5791     3526       GenBank         complete      NaN     NaN  ...   \n",
       "2486     4215        RefSeq         complete      NaN     NaN  ...   \n",
       "11811    4218       GenBank         complete      NaN     NaN  ...   \n",
       "11813    4218       GenBank         complete      NaN     NaN  ...   \n",
       "11814    4218       GenBank         complete      NaN     NaN  ...   \n",
       "11815    4219       GenBank         complete      NaN     NaN  ...   \n",
       "11812    4221       GenBank         complete      NaN     NaN  ...   \n",
       "11816    4241       GenBank         complete      NaN     NaN  ...   \n",
       "11817    4243       GenBank         complete      NaN     NaN  ...   \n",
       "11819    4273       GenBank         complete      NaN     NaN  ...   \n",
       "11818    4273       GenBank         complete      NaN     NaN  ...   \n",
       "4462     5342       GenBank         complete      NaN     NaN  ...   \n",
       "11577    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11983    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11578    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11586    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11576    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11575    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11574    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11569    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11568    5386       GenBank         complete      NaN     NaN  ...   \n",
       "11989    5386       GenBank         complete      NaN     NaN  ...   \n",
       "10506    5386       GenBank         complete      NaN     NaN  ...   \n",
       "\n",
       "      Host_phylum_id Host_phylum_name Host_class_id      Host_class_name  \\\n",
       "11809         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11098         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "1711          1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11808         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11810         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "2512          1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11173         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "5791          1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "2486          1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11811         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11813         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11814         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11815         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11812         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11816         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11817         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11819         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11818         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "4462          1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11577         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11983         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11578         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11586         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11576         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11575         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11574         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11569         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11568         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "11989         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "10506         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "\n",
       "      Host_order_id   Host_order_name Host_family_id    Host_family_name  \\\n",
       "11809       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11098       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "1711        91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11808       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11810       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "2512        91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11173       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "5791        91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "2486        91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11811       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11813       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11814       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11815       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11812       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11816       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11817       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11819       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11818       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "4462        91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11577       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11983       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11578       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11586       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11576       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11575       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11574       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11569       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11568       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "11989       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "10506       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "\n",
       "       Host_species_id  Host_species_name  \n",
       "11809              562   Escherichia coli  \n",
       "11098              562   Escherichia coli  \n",
       "1711               562   Escherichia coli  \n",
       "11808              562   Escherichia coli  \n",
       "11810              562   Escherichia coli  \n",
       "2512               562   Escherichia coli  \n",
       "11173              562   Escherichia coli  \n",
       "5791               562   Escherichia coli  \n",
       "2486               562   Escherichia coli  \n",
       "11811              562   Escherichia coli  \n",
       "11813              562   Escherichia coli  \n",
       "11814              562   Escherichia coli  \n",
       "11815              562   Escherichia coli  \n",
       "11812              562   Escherichia coli  \n",
       "11816              562   Escherichia coli  \n",
       "11817              562   Escherichia coli  \n",
       "11819              562   Escherichia coli  \n",
       "11818              562   Escherichia coli  \n",
       "4462               562   Escherichia coli  \n",
       "11577              562   Escherichia coli  \n",
       "11983              562   Escherichia coli  \n",
       "11578              562   Escherichia coli  \n",
       "11586              562   Escherichia coli  \n",
       "11576              562   Escherichia coli  \n",
       "11575              562   Escherichia coli  \n",
       "11574              562   Escherichia coli  \n",
       "11569              562   Escherichia coli  \n",
       "11568              562   Escherichia coli  \n",
       "11989              562   Escherichia coli  \n",
       "10506              562   Escherichia coli  \n",
       "\n",
       "[30 rows x 33 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df.sort_values('Length').head(n=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "138px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
