{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO, Entrez\n",
    "Entrez.email = 'adam.hockenberry@utexas.edu'  # Always tell NCBI who you are\n",
    "import time\n",
    "import glob"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# First make the starting datatable better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17345, 18)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Publications</th>\n",
       "      <th>Geo_Location</th>\n",
       "      <th>Host</th>\n",
       "      <th>Isolation_Source</th>\n",
       "      <th>Collection_Date</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>GenBank_Title</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NC_045425</td>\n",
       "      <td>2019-12-16T00:00:00Z</td>\n",
       "      <td>Thermus virus OH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>5688</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nagayoshi,Y., Kumagae,K., Mori,K., Tashiro,K.,...</td>\n",
       "      <td>26941711</td>\n",
       "      <td>Japan</td>\n",
       "      <td>Thermus thermophilus HB8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Thermus phage phiOH3 genomic DNA, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC_044940</td>\n",
       "      <td>2019-10-02T00:00:00Z</td>\n",
       "      <td>Pectinobacterium virus PEAT2</td>\n",
       "      <td>Peatvirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>48659</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Kalischuk,M., Hachey,J., Thomas,D., Kawchuk,L.</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Pectobacterium phage PEAT2, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC_043767</td>\n",
       "      <td>2019-07-20T00:00:00Z</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "      <td>Rosebushvirus</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>67324</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Lunt,B.L., Payne,D.E., Fisher,J.N.B., Smith,K....</td>\n",
       "      <td>NaN</td>\n",
       "      <td>USA: Provo, UT</td>\n",
       "      <td>Mycolicibacterium smegmatis MC2 155</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2009-09-13</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Mycobacterium virus TA17a, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NC_043027</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Bacillus virus PBS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>252197</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Russell,D.A., Jacobs-Sera,D., Duda,R., Hatfull...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Bacillus virus PBS1, complete genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NC_043028</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Xanthomonas virus Xf109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7190</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yeh,T.Y.</td>\n",
       "      <td>27743252</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xanthomonas oryzae ATCC 35933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Xanthomonas phage Xf109, complete genome</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession          Release_Date                       Species  \\\n",
       "0  NC_045425  2019-12-16T00:00:00Z             Thermus virus OH3   \n",
       "1  NC_044940  2019-10-02T00:00:00Z  Pectinobacterium virus PEAT2   \n",
       "2  NC_043767  2019-07-20T00:00:00Z     Mycobacterium virus TA17a   \n",
       "3  NC_043027  2019-06-28T00:00:00Z           Bacillus virus PBS1   \n",
       "4  NC_043028  2019-06-28T00:00:00Z       Xanthomonas virus Xf109   \n",
       "\n",
       "           Genus        Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "0            NaN    Inoviridae    5688        RefSeq         complete   \n",
       "1      Peatvirus    Myoviridae   48659        RefSeq         complete   \n",
       "2  Rosebushvirus  Siphoviridae   67324        RefSeq         complete   \n",
       "3            NaN    Myoviridae  252197        RefSeq         complete   \n",
       "4            NaN    Inoviridae    7190        RefSeq         complete   \n",
       "\n",
       "  Genotype Segment                                            Authors  \\\n",
       "0      NaN     NaN  Nagayoshi,Y., Kumagae,K., Mori,K., Tashiro,K.,...   \n",
       "1      NaN     NaN     Kalischuk,M., Hachey,J., Thomas,D., Kawchuk,L.   \n",
       "2      NaN     NaN  Lunt,B.L., Payne,D.E., Fisher,J.N.B., Smith,K....   \n",
       "3      NaN     NaN  Russell,D.A., Jacobs-Sera,D., Duda,R., Hatfull...   \n",
       "4      NaN     NaN                                           Yeh,T.Y.   \n",
       "\n",
       "  Publications    Geo_Location                                 Host  \\\n",
       "0     26941711           Japan             Thermus thermophilus HB8   \n",
       "1          NaN             NaN                                  NaN   \n",
       "2          NaN  USA: Provo, UT  Mycolicibacterium smegmatis MC2 155   \n",
       "3          NaN             NaN                    Bacillus subtilis   \n",
       "4     27743252             NaN        Xanthomonas oryzae ATCC 35933   \n",
       "\n",
       "  Isolation_Source Collection_Date BioSample  \\\n",
       "0              NaN             NaN       NaN   \n",
       "1              NaN             NaN       NaN   \n",
       "2              NaN      2009-09-13       NaN   \n",
       "3              NaN             NaN       NaN   \n",
       "4              NaN             NaN       NaN   \n",
       "\n",
       "                                       GenBank_Title  \n",
       "0  Thermus phage phiOH3 genomic DNA, complete genome  \n",
       "1        Pectobacterium phage PEAT2, complete genome  \n",
       "2         Mycobacterium virus TA17a, complete genome  \n",
       "3               Bacillus virus PBS1, complete genome  \n",
       "4           Xanthomonas phage Xf109, complete genome  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../Data/NCBI_phage_db/all_complete_phage_info.csv')\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12483, 18)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['Host'].isnull()==False]\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mycolicibacterium smegmatis MC2 155        1727\n",
       "Escherichia coli                            745\n",
       "Gordonia terrae                             382\n",
       "Escherichia coli C                          346\n",
       "Lactococcus lactis                          323\n",
       "                                           ... \n",
       "Streptococcus pseudoporcinus SPIN 20026       1\n",
       "Streptococcus pneumoniae GA41317              1\n",
       "Streptococcus uberis C5388                    1\n",
       "Escherichia coli PA8                          1\n",
       "Leptospira                                    1\n",
       "Name: Host, Length: 967, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Better treatment of host taxonomy**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H43 str. T22\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O145:NM\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Nodularia\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Gordonia\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O104:H4\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli BL21(DE3)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Bacillus\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H7 str. EDL933\n",
      "Skipping\n",
      "[Bacillus] clarkii\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Yersinia enterocolitica (type O:3)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Streptomyces coelicolor A3(2)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Edwardsiella\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "[Brevibacterium] flavum\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Curvibacter sp. AEP1.3\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Streptococcus anginosus MAS624\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Morganella sp. (in: Bacteria)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Clostridium saccharoperbutylacetonicum N1-4(HMT)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Skipping\n",
      "Erwinia horticola\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O157:H7\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O111:H2\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli O103:H25 str. NIPH-11060424\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Escherichia coli DH5[alpha]\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Bacillus sp. (in: Bacteria)\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Vibrio maritimus\n",
      "Vibrio campbellii ATCC BAA-1116\n",
      "Flavobacterium sp. LMO6\n",
      "Streptococcus\n",
      "Geobacillus stearothermophilus\n",
      "Streptococcus suis TL13\n",
      "Brucella abortus 544\n",
      "Corynebacterium glutamicum\n",
      "Staphylococcus hominis\n",
      "Streptococcus infantarius\n",
      "Aeromonas media\n",
      "Rheinheimera sp. BAL341\n",
      "Escherichia coli PA2\n",
      "Streptococcus henryi DSM 19005\n",
      "Cronobacter sakazakii\n",
      "Streptomyces azureus\n",
      "Streptococcus agalactiae ILRI005\n",
      "Burkholderia cepacia\n",
      "Streptococcus agalactiae GD201008-001\n",
      "Sinorhizobium meliloti 1021\n",
      "Pseudoalteromonas hodoensis\n",
      "Vibrio coralliilyticus\n",
      "Escherichia sp. MOD1-EC6098\n",
      "Ochrobactrum sp. POC9\n",
      "Streptococcus thermophilus ND03\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Lysinibacillus sphaericus C3-41\n",
      "Thermus sp. TC4\n",
      "Lactococcus garvieae Lg2\n",
      "Pseudomonas syringae\n",
      "Morganella morganii\n",
      "Microcystis aeruginosa\n",
      "Enterococcus faecalis 62\n",
      "Synechococcus sp. WH 8018\n",
      "Fusobacterium nucleatum\n",
      "Streptococcus intermedius\n",
      "Vibrio cholerae O1\n",
      "Streptococcus anginosus\n",
      "Sphaerotilus natans\n",
      "Idiomarinaceae bacterium N2-2\n",
      "Vibrio metschnikovii\n",
      "Streptococcus pneumoniae BS458\n",
      "Acaryochloris marina\n",
      "Rhizobium gallicum\n",
      "Acinetobacter baumannii A118\n",
      "Streptomyces toxytricini\n",
      "Sodalis glossinidius\n",
      "Mycolicibacterium aichiense\n",
      "Escherichia coli C\n",
      "Streptococcus pneumoniae ST556\n",
      "Lactococcus lactis subsp. cremoris TIFN7\n",
      "Streptococcus pneumoniae GA47033\n",
      "Bacteroides intestinalis\n",
      "Polaribacter\n",
      "Streptomyces sp.\n",
      "Lactobacillus\n",
      "Epibacterium mobile\n",
      "Microbacterium natoriense\n",
      "Caulobacter vibrioides CB15\n",
      "Streptococcus equinus\n",
      "Trichormus variabilis\n",
      "Plesiomonas sp.\n",
      "Streptococcus parauberis KCTC 11980BP\n",
      "Streptococcus suis ST1\n",
      "Aeromonas rivipollensis\n",
      "Pseudomonas aeruginosa CHA\n",
      "Myxococcus xanthus\n",
      "Sphingomonas paucimobilis\n",
      "Escherichia coli H8\n",
      "Clostridioides difficile\n",
      "Parabacteroides merdae\n",
      "Streptococcus pneumoniae GA41538\n",
      "Thermobifida fusca\n",
      "Gordonia alkanivorans\n",
      "Prochlorococcus marinus\n",
      "Streptococcus gordonii\n",
      "Yersinia\n",
      "Marinomonas sp.\n",
      "Sphingobium sp. IP1\n",
      "Halomonas ventosae\n",
      "Mycobacteroides abscessus subsp. bolletii F1725\n",
      "Escherichia coli J53\n",
      "Streptococcus parauberis KRS-02083\n",
      "Nonlabens\n",
      "Streptococcus mitis SK564\n",
      "Tsukamurella paurometabola\n",
      "Shigella dysenteriae\n",
      "Streptococcus suis YB51\n",
      "Enterococcus\n",
      "Cellulophaga baltica NN016038\n",
      "Streptomyces virginiae\n",
      "Streptomyces avermitilis\n",
      "Dinoroseobacter shibae\n",
      "Campylobacter jejuni subsp. jejuni\n",
      "Streptococcus agalactiae BSU450\n",
      "Phormidium sp. MIS-PhA\n",
      "Parabacteroides distasonis\n",
      "Cronobacter\n",
      "Clostridium perfringens ATCC 13124\n",
      "Tenacibaculum geojense\n",
      "Staphylococcus aureus subsp. aureus N315\n",
      "Pseudomonas fluorescens group\n",
      "Burkholderia pseudomallei K96243\n",
      "Paracoccus marcusii\n",
      "Streptococcus suis 98HAH33\n",
      "Staphylococcus haemolyticus\n",
      "Lactococcus\n",
      "Pseudoalteromonas marina\n",
      "Brevundimonas subvibrioides ATCC 15264\n",
      "Erwinia pyrifoliae\n",
      "Alteromonas\n",
      "Salmonella enterica subsp. enterica serovar Typhimurium str. LT2\n",
      "Paracoccus sulfuroxidans\n",
      "Mesorhizobium loti\n",
      "Propionibacterium\n",
      "Erysipelothrix rhusiopathiae\n",
      "Aminobacter sp. MSH1\n",
      "Shewanella inventionis\n",
      "Rhodoferax sp. IMCC26218\n",
      "Rhodobacter\n",
      "Bacillus pumilus BL8\n",
      "Clostridium tetani\n",
      "Lactobacillus delbrueckii subsp. bulgaricus\n",
      "Salinibacter ruber\n",
      "Microbacterium foliorum\n",
      "Pseudomonas aeruginosa PA14\n",
      "Streptomyces tricolor\n",
      "Streptococcus anginosus F0211\n",
      "Synechococcus sp. WH 8102\n",
      "Stenotrophomonas sp.\n",
      "Haemophilus influenzae\n",
      "Streptococcus pyogenes NZ131\n",
      "Paracoccus\n",
      "Streptococcus pseudopneumoniae 5247\n",
      "Streptococcus pseudopneumoniae 22725\n",
      "Salmonella enterica subsp. enterica serovar Rissen\n",
      "Streptococcus uberis B190\n",
      "Vibrio sp.\n",
      "Salmonella enterica subsp. enterica serovar Saintpaul\n",
      "Streptococcus pneumoniae GA17328\n",
      "Streptococcus oralis ATCC 49296\n",
      "Oncorhynchus mykiss\n",
      "Shewanella fidelis\n",
      "Gordonia sp.\n",
      "Streptococcus agalactiae GB00640\n",
      "Lactobacillus plantarum\n",
      "Vibrio mediterranei\n",
      "Prochlorococcus marinus str. MIT 9313\n",
      "Pseudomonas chlororaphis\n",
      "Streptococcus hyovaginalis DSM 12219\n",
      "Synechococcus sp. WH 7803\n",
      "Streptococcus thermophilus 1F8CT\n",
      "Aeromonas salmonicida subsp. salmonicida\n",
      "Mycobacterium sp.\n",
      "Pectobacterium carotovorum\n",
      "Streptococcus parauberis KCTC 11537\n",
      "Photobacterium damselae\n",
      "Bacillus subtilis\n",
      "Bacillus thuringiensis serovar kurstaki\n",
      "Aggregatibacter actinomycetemcomitans D11S-1\n",
      "Bacillus amyloliquefaciens\n",
      "Acinetobacter johnsonii\n",
      "Candidatus Pelagibacter\n",
      "Streptococcus pneumoniae 23F\n",
      "Salinivibrio sp.\n",
      "Providencia sp.\n",
      "Streptococcus phocae subsp. salmonis\n",
      "Escherichia coli PA33\n",
      "Acholeplasma laidlawii\n",
      "Microbacterium nematophilum\n",
      "Achromobacter animicus\n",
      "Shigella\n",
      "Escherichia coli PA11\n"
     ]
    }
   ],
   "source": [
    "taxonomies_to_fetch = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "problematic_hosts = []\n",
    "for host in list(set(df['Host']))[:]:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) != 1:\n",
    "        problematic_hosts.append(host)\n",
    "        continue\n",
    "    record_id = record['IdList'][0]\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        problematic_hosts.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Mycolicibacterium smegmatis    1978\n",
       "Escherichia coli               1187\n",
       "Salmonella enterica             423\n",
       "Pseudomonas aeruginosa          400\n",
       "Gordonia terrae                 382\n",
       "                               ... \n",
       "Salipiger profundus               1\n",
       "Pseudomonas sp. 1-1-1b            1\n",
       "Rothia dentocariosa               1\n",
       "Phaeobacter inhibens              1\n",
       "Microcystis elabens               1\n",
       "Name: Host_species_name, Length: 549, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_species_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11    12021\n",
       "1       270\n",
       "4        15\n",
       "Name: GeneticCode, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['GeneticCode'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Fix the problematic hosts**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "27\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Escherichia coli O157:H43 str. T22',\n",
       " 'Escherichia coli O145:NM',\n",
       " 'Nodularia',\n",
       " 'Gordonia',\n",
       " 'Escherichia coli O104:H4',\n",
       " 'Escherichia coli BL21(DE3)',\n",
       " 'Bacillus',\n",
       " 'Escherichia coli O157:H7 str. EDL933',\n",
       " '[Bacillus] clarkii',\n",
       " 'Yersinia enterocolitica (type O:3)',\n",
       " 'Streptomyces coelicolor A3(2)',\n",
       " 'Edwardsiella',\n",
       " '[Brevibacterium] flavum',\n",
       " 'Curvibacter sp. AEP1.3',\n",
       " 'Streptococcus anginosus MAS624',\n",
       " 'Morganella sp. (in: Bacteria)',\n",
       " 'Clostridium saccharoperbutylacetonicum N1-4(HMT)',\n",
       " 'Methylophilaceae bacterium IMCC19250',\n",
       " 'Erwinia horticola',\n",
       " 'Escherichia coli O157:H7',\n",
       " 'Escherichia coli O111:H2',\n",
       " 'Escherichia coli O103:H25 str. NIPH-11060424',\n",
       " 'Escherichia coli DH5[alpha]',\n",
       " 'Bacillus sp. (in: Bacteria)',\n",
       " 'Yersinia',\n",
       " 'Phormidium sp. MIS-PhA',\n",
       " 'Paracoccus']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(set(problematic_hosts)))\n",
    "problematic_hosts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Cases with special characters**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Escherichia coli O157 H43 str. T22\n",
      "Escherichia coli O145 NM\n",
      "Nodularia\n",
      "Gordonia\n",
      "Escherichia coli O104 H4\n",
      "Escherichia coli BL21 DE3 \n",
      "Bacillus\n",
      "Escherichia coli O157 H7 str. EDL933\n",
      "Bacillus clarkii\n",
      "Yersinia enterocolitica  type O 3 \n",
      "Streptomyces coelicolor A3 2 \n",
      "Edwardsiella\n",
      "Brevibacterium flavum\n",
      "Curvibacter sp. AEP1.3\n",
      "Streptococcus anginosus MAS624\n",
      "Morganella sp.  in  Bacteria \n",
      "Clostridium saccharoperbutylacetonicum N1-4 HMT \n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Erwinia horticola\n",
      "Escherichia coli O157 H7\n",
      "Escherichia coli O111 H2\n",
      "Escherichia coli O103 H25 str. NIPH-11060424\n",
      "Escherichia coli DH5alpha\n",
      "Bacillus sp.  in  Bacteria \n",
      "Yersinia\n",
      "Phormidium sp. MIS-PhA\n",
      "Paracoccus\n"
     ]
    }
   ],
   "source": [
    "still_problematic = []\n",
    "for host in problematic_hosts:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    host = host.replace(':', ' ')\n",
    "    host = host.replace('[', '').replace(']', '')\n",
    "    host = host.replace('(', ' ').replace(')', ' ')\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) != 1:\n",
    "        still_problematic.append(host)\n",
    "        continue\n",
    "    record_id = record['IdList'][0]\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        still_problematic.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['Nodularia',\n",
       " 'Gordonia',\n",
       " 'Bacillus',\n",
       " 'Edwardsiella',\n",
       " 'Curvibacter sp. AEP1.3',\n",
       " 'Streptococcus anginosus MAS624',\n",
       " 'Methylophilaceae bacterium IMCC19250',\n",
       " 'Erwinia horticola',\n",
       " 'Yersinia',\n",
       " 'Phormidium sp. MIS-PhA',\n",
       " 'Paracoccus']"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(set(still_problematic)))\n",
    "still_problematic"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Cases with annoying genuses**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodularia\n",
      "Gordonia\n",
      "Bacillus\n",
      "Edwardsiella\n",
      "Curvibacter sp. AEP1.3\n",
      "Streptococcus anginosus MAS624\n",
      "Methylophilaceae bacterium IMCC19250\n",
      "Erwinia horticola\n",
      "Yersinia\n",
      "Phormidium sp. MIS-PhA\n",
      "Paracoccus\n"
     ]
    }
   ],
   "source": [
    "really_problematic = []\n",
    "for host in still_problematic:\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    host = host.replace(':', ' ')\n",
    "    host = host.replace('[', '').replace(']', '')\n",
    "    host = host.replace('(', ' ').replace(')', ' ')\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    handle = Entrez.esearch(db='Taxonomy', term=host)\n",
    "    record = Entrez.read(handle)\n",
    "    if len(record['IdList']) == 0:\n",
    "        really_problematic.append(host)\n",
    "        continue\n",
    "    temp_ids = record['IdList']\n",
    "    successful_ids = []\n",
    "    for record_id in temp_ids:\n",
    "        handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "        records = Entrez.read(handle)\n",
    "        if len(records) != 1:\n",
    "            really_problematic.append(host)\n",
    "            continue        \n",
    "        record = records[0]\n",
    "        lineage = record['LineageEx']\n",
    "        for i in lineage:\n",
    "            if i['Rank']=='superkingdom' and i['ScientificName'] =='Bacteria':\n",
    "                successful_ids.append(record_id)\n",
    "    if len(successful_ids) == 1:\n",
    "        record_id = successful_ids[0]\n",
    "    else:\n",
    "        really_problematic.append(host)\n",
    "    \n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)\n",
    "    if len(records) != 1:\n",
    "        really_problematic.append(host)\n",
    "        continue        \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5\n",
      "['Curvibacter sp. AEP1.3', 'Streptococcus anginosus MAS624', 'Methylophilaceae bacterium IMCC19250', 'Erwinia horticola', 'Phormidium sp. MIS-PhA']\n"
     ]
    }
   ],
   "source": [
    "print(len(really_problematic))\n",
    "print(really_problematic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "answers = [2685271, 1328, 2030816, 551, 1199]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n",
      "Skipping\n"
     ]
    }
   ],
   "source": [
    "taxonomies_to_fetch = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "problematic_hosts = []\n",
    "for host, answer in zip(really_problematic, answers):\n",
    "    matching_indices = df[df['Host']==host].index\n",
    "    if df.loc[matching_indices]['Host_superkingdom_id'].isnull().all()==False:\n",
    "        print('Skipping')\n",
    "        continue\n",
    "    print(host)\n",
    "    record_id = str(answer)\n",
    "    handle = Entrez.efetch(db=\"Taxonomy\", id=record_id, retmode=\"xml\")\n",
    "    records = Entrez.read(handle)    \n",
    "    record = records[0]\n",
    "    ###First get the genetic code right\n",
    "    df.at[matching_indices, 'GeneticCode'] = record['GeneticCode']['GCId']\n",
    "    ###Now possibly add the record itself\n",
    "    if record ['Rank'] in taxonomies_to_fetch:\n",
    "        df.at[matching_indices, 'Host_{}_id'.format(record['Rank'])] = record['TaxId']\n",
    "        df.at[matching_indices, 'Host_{}_name'.format(record['Rank'])] = record['ScientificName']\n",
    "    ###Finally go through the lineage\n",
    "    lineage = record['LineageEx']\n",
    "    for i in lineage:\n",
    "        if i['Rank'] in taxonomies_to_fetch:\n",
    "            df.at[matching_indices, 'Host_{}_id'.format(i['Rank'])] = i['TaxId']\n",
    "            df.at[matching_indices, 'Host_{}_name'.format(i['Rank'])] = i['ScientificName']\n",
    "    time.sleep(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Write the new and improved file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Bacteria     12211\n",
       "Eukaryota      270\n",
       "Archaea          2\n",
       "Name: Host_superkingdom_name, dtype: int64"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_superkingdom_name'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12211, 33)\n"
     ]
    }
   ],
   "source": [
    "df = df[df['Host_superkingdom_name']=='Bacteria']\n",
    "print(df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../Data/NCBI_phage_db/all_complete_phage_info_HOSTTAXONOMY_05152020.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select the rows that I care about"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12211, 33)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>Segment</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_species_id</th>\n",
       "      <th>Host_species_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>NC_045425</td>\n",
       "      <td>2019-12-16T00:00:00Z</td>\n",
       "      <td>Thermus virus OH3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>5688</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1297.0</td>\n",
       "      <td>Deinococcus-Thermus</td>\n",
       "      <td>188787.0</td>\n",
       "      <td>Deinococci</td>\n",
       "      <td>68933.0</td>\n",
       "      <td>Thermales</td>\n",
       "      <td>188786.0</td>\n",
       "      <td>Thermaceae</td>\n",
       "      <td>274.0</td>\n",
       "      <td>Thermus thermophilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>NC_043767</td>\n",
       "      <td>2019-07-20T00:00:00Z</td>\n",
       "      <td>Mycobacterium virus TA17a</td>\n",
       "      <td>Rosebushvirus</td>\n",
       "      <td>Siphoviridae</td>\n",
       "      <td>67324</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>201174.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>1760.0</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>85007.0</td>\n",
       "      <td>Corynebacteriales</td>\n",
       "      <td>1762.0</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>1772.0</td>\n",
       "      <td>Mycolicibacterium smegmatis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>NC_043027</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Bacillus virus PBS1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>252197</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1239.0</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>91061.0</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>1385.0</td>\n",
       "      <td>Bacillales</td>\n",
       "      <td>186817.0</td>\n",
       "      <td>Bacillaceae</td>\n",
       "      <td>1423.0</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>NC_043028</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Xanthomonas virus Xf109</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7190</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>347.0</td>\n",
       "      <td>Xanthomonas oryzae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>NC_043029</td>\n",
       "      <td>2019-06-28T00:00:00Z</td>\n",
       "      <td>Stenotrophomonas virus SMA6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inoviridae</td>\n",
       "      <td>7648</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>135614.0</td>\n",
       "      <td>Xanthomonadales</td>\n",
       "      <td>32033.0</td>\n",
       "      <td>Xanthomonadaceae</td>\n",
       "      <td>40324.0</td>\n",
       "      <td>Stenotrophomonas maltophilia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession          Release_Date                      Species  \\\n",
       "0  NC_045425  2019-12-16T00:00:00Z            Thermus virus OH3   \n",
       "1  NC_043767  2019-07-20T00:00:00Z    Mycobacterium virus TA17a   \n",
       "2  NC_043027  2019-06-28T00:00:00Z          Bacillus virus PBS1   \n",
       "3  NC_043028  2019-06-28T00:00:00Z      Xanthomonas virus Xf109   \n",
       "4  NC_043029  2019-06-28T00:00:00Z  Stenotrophomonas virus SMA6   \n",
       "\n",
       "           Genus        Family  Length Sequence_Type Nuc_Completeness  \\\n",
       "0            NaN    Inoviridae    5688        RefSeq         complete   \n",
       "1  Rosebushvirus  Siphoviridae   67324        RefSeq         complete   \n",
       "2            NaN    Myoviridae  252197        RefSeq         complete   \n",
       "3            NaN    Inoviridae    7190        RefSeq         complete   \n",
       "4            NaN    Inoviridae    7648        RefSeq         complete   \n",
       "\n",
       "  Genotype Segment  ... Host_phylum_id     Host_phylum_name Host_class_id  \\\n",
       "0      NaN     NaN  ...         1297.0  Deinococcus-Thermus      188787.0   \n",
       "1      NaN     NaN  ...       201174.0       Actinobacteria        1760.0   \n",
       "2      NaN     NaN  ...         1239.0           Firmicutes       91061.0   \n",
       "3      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "4      NaN     NaN  ...         1224.0       Proteobacteria        1236.0   \n",
       "\n",
       "       Host_class_name Host_order_id    Host_order_name Host_family_id  \\\n",
       "0           Deinococci       68933.0          Thermales       188786.0   \n",
       "1       Actinobacteria       85007.0  Corynebacteriales         1762.0   \n",
       "2              Bacilli        1385.0         Bacillales       186817.0   \n",
       "3  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "4  Gammaproteobacteria      135614.0    Xanthomonadales        32033.0   \n",
       "\n",
       "   Host_family_name  Host_species_id             Host_species_name  \n",
       "0        Thermaceae            274.0          Thermus thermophilus  \n",
       "1  Mycobacteriaceae           1772.0   Mycolicibacterium smegmatis  \n",
       "2       Bacillaceae           1423.0             Bacillus subtilis  \n",
       "3  Xanthomonadaceae            347.0            Xanthomonas oryzae  \n",
       "4  Xanthomonadaceae          40324.0  Stenotrophomonas maltophilia  \n",
       "\n",
       "[5 rows x 33 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('../Data/NCBI_phage_db/all_complete_phage_info_HOSTTAXONOMY_05152020.tsv', sep='\\t')\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Enterobacteriaceae      2453\n",
       "Mycobacteriaceae        2143\n",
       "Streptococcaceae        1446\n",
       "Pseudomonadaceae         617\n",
       "Gordoniaceae             452\n",
       "Synechococcaceae         386\n",
       "Bacillaceae              384\n",
       "Staphylococcaceae        351\n",
       "Vibrionaceae             326\n",
       "Micrococcaceae           304\n",
       "Streptomycetaceae        254\n",
       "Microbacteriaceae        253\n",
       "Propionibacteriaceae     211\n",
       "Moraxellaceae            182\n",
       "Flavobacteriaceae        181\n",
       "Lactobacillaceae         128\n",
       "Enterococcaceae          126\n",
       "Burkholderiaceae         122\n",
       "Pectobacteriaceae        105\n",
       "Erwiniaceae              105\n",
       "Name: Host_family_name, dtype: int64"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['Host_family_name'].value_counts()[:20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_accession_numbers = ['NC_001604', 'NC_001416']\n",
    "all_genomes = SeqIO.parse('../Data/NCBI_phage_db/all_complete_phage_nts.fasta', 'fasta')\n",
    "for genome in all_genomes:\n",
    "    if genome.id in selected_accession_numbers:\n",
    "        with open('../Data/NCBI_phage_db/phage_genomes/{}.fasta'.format(genome.id), 'w') as outfile:\n",
    "            SeqIO.write(genome, outfile, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Split apart the combined genome `fasta` file to write the genomes that I selected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = SeqIO.read('../Data/NCBI_phage_db/phage_genomes/NC_001416.fasta', 'fasta')\n",
    "# for genome_file in glob.glob('../Data/NCBI_phage_db/phage_genomes/*.fasta'):\n",
    "#     genome = SeqIO.read(genome_file, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = []\n",
    "all_cds = SeqIO.parse('../Data/NCBI_phage_db/all_complete_phage_CDS.fasta', 'fasta')\n",
    "for cds in all_cds:\n",
    "    if genome.id in cds.id.split(':')[0]:\n",
    "        hits.append(cds)\n",
    "print(len(hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(hits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for hit in hits:\n",
    "    reverse_complement = False\n",
    "    if 'join' in hit.id:\n",
    "        continue\n",
    "    if 'complement(' == hit.id[:11]:\n",
    "        print('here')\n",
    "        reverse_complement = True\n",
    "#         hit.id = hit.id[11:-1]\n",
    "    if len(hit.id.split(':')) == 2:\n",
    "        region = hit.id.split(':')[-1]\n",
    "        if len(region.split('..')) == 2:\n",
    "            start = int(region.split('..')[0])\n",
    "            stop = int(region.split('..')[-1])\n",
    "    print(start, stop)\n",
    "    print(genome.seq[start-1:stop]==hit.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make CDS datatables for each of those genomes by merging the genome `fasta` file with the CDS `fasta` file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**et voila**"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "138px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
