{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Offline portion\n",
    "\n",
    "I manually ran fastANI on the set of phage genomes (`.fasta` files) for each host.\n",
    "\n",
    "For example:\n",
    "```\n",
    "find ./phage_genomes/562_phage_genomes/*.fasta -type f> file_listing.txt\n",
    "\n",
    "fastANI --ql file_listing.txt --rl file_listing.txt --threads 10 --fragLen 300 --minFraction 0.8 -o 562_ANI.output\n",
    "```\n",
    "\n",
    "To run for all hosts, just replace both instances of `562` in the above code. Didn't seem worth writing a bash script for. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Now using that information to add a new column to the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12876, 35)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Accession</th>\n",
       "      <th>SRA_Accession</th>\n",
       "      <th>Release_Date</th>\n",
       "      <th>Species</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Family</th>\n",
       "      <th>Length</th>\n",
       "      <th>Sequence_Type</th>\n",
       "      <th>Nuc_Completeness</th>\n",
       "      <th>Genotype</th>\n",
       "      <th>...</th>\n",
       "      <th>Host_phylum_id</th>\n",
       "      <th>Host_phylum_name</th>\n",
       "      <th>Host_class_id</th>\n",
       "      <th>Host_class_name</th>\n",
       "      <th>Host_order_id</th>\n",
       "      <th>Host_order_name</th>\n",
       "      <th>Host_family_id</th>\n",
       "      <th>Host_family_name</th>\n",
       "      <th>Host_genus_id</th>\n",
       "      <th>Host_genus_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_050154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-13T00:00:00Z</td>\n",
       "      <td>Escherichia phage D6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>91159</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>91347.0</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>543.0</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>561.0</td>\n",
       "      <td>Escherichia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_050143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage datas</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>60746</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>72274.0</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>135621.0</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>286.0</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_050144</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage Epa14</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>65797</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>72274.0</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>135621.0</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>286.0</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_050145</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage PaGU11</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>65554</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>72274.0</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>135621.0</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>286.0</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NC_050146</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2020-08-10T00:00:00Z</td>\n",
       "      <td>Pseudomonas phage Epa7</td>\n",
       "      <td>Pbunavirus</td>\n",
       "      <td>Myoviridae</td>\n",
       "      <td>65629</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>complete</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1224.0</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>1236.0</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>72274.0</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>135621.0</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>286.0</td>\n",
       "      <td>Pseudomonas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Accession SRA_Accession          Release_Date                   Species  \\\n",
       "0  NC_050154           NaN  2020-08-13T00:00:00Z      Escherichia phage D6   \n",
       "1  NC_050143           NaN  2020-08-10T00:00:00Z   Pseudomonas phage datas   \n",
       "2  NC_050144           NaN  2020-08-10T00:00:00Z   Pseudomonas phage Epa14   \n",
       "3  NC_050145           NaN  2020-08-10T00:00:00Z  Pseudomonas phage PaGU11   \n",
       "4  NC_050146           NaN  2020-08-10T00:00:00Z    Pseudomonas phage Epa7   \n",
       "\n",
       "        Genus      Family  Length Sequence_Type Nuc_Completeness Genotype  \\\n",
       "0         NaN  Myoviridae   91159        RefSeq         complete      NaN   \n",
       "1  Pbunavirus  Myoviridae   60746        RefSeq         complete      NaN   \n",
       "2  Pbunavirus  Myoviridae   65797        RefSeq         complete      NaN   \n",
       "3  Pbunavirus  Myoviridae   65554        RefSeq         complete      NaN   \n",
       "4  Pbunavirus  Myoviridae   65629        RefSeq         complete      NaN   \n",
       "\n",
       "   ... Host_phylum_id Host_phylum_name Host_class_id      Host_class_name  \\\n",
       "0  ...         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "1  ...         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "2  ...         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "3  ...         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "4  ...         1224.0   Proteobacteria        1236.0  Gammaproteobacteria   \n",
       "\n",
       "  Host_order_id   Host_order_name Host_family_id    Host_family_name  \\\n",
       "0       91347.0  Enterobacterales          543.0  Enterobacteriaceae   \n",
       "1       72274.0   Pseudomonadales       135621.0    Pseudomonadaceae   \n",
       "2       72274.0   Pseudomonadales       135621.0    Pseudomonadaceae   \n",
       "3       72274.0   Pseudomonadales       135621.0    Pseudomonadaceae   \n",
       "4       72274.0   Pseudomonadales       135621.0    Pseudomonadaceae   \n",
       "\n",
       "  Host_genus_id Host_genus_name  \n",
       "0         561.0     Escherichia  \n",
       "1         286.0     Pseudomonas  \n",
       "2         286.0     Pseudomonas  \n",
       "3         286.0     Pseudomonas  \n",
       "4         286.0     Pseudomonas  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_df = pd.read_csv('../Data/Other_possible_dbs/NCBI_phage_db/paper_dataset_11_2020.tsv', sep='\\t')\n",
    "print(full_df.shape)\n",
    "full_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n",
      "[1772, 562, 28901, 2055, 1358, 287, 573, 104336, 1280, 1771959]\n"
     ]
    }
   ],
   "source": [
    "vc = full_df['Host_species_id'].value_counts()\n",
    "taxonomy_list = list(vc[vc >= 50].index)\n",
    "print(len(taxonomy_list))\n",
    "print(taxonomy_list[:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Looking at individual dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (148432, 5)\n",
      "Non-null shape: (148432, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>./phage_genomes/562_phage_genomes/AF063097.fasta</td>\n",
       "      <td>./phage_genomes/562_phage_genomes/AF063097.fasta</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>./phage_genomes/562_phage_genomes/AF063097.fasta</td>\n",
       "      <td>./phage_genomes/562_phage_genomes/NC_001895.fasta</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>./phage_genomes/562_phage_genomes/AP019559.fasta</td>\n",
       "      <td>./phage_genomes/562_phage_genomes/NC_048627.fasta</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>369</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>./phage_genomes/562_phage_genomes/AP019559.fasta</td>\n",
       "      <td>./phage_genomes/562_phage_genomes/AP019559.fasta</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>369</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>./phage_genomes/562_phage_genomes/AP019559.fasta</td>\n",
       "      <td>./phage_genomes/562_phage_genomes/KJ190157.fasta</td>\n",
       "      <td>95.7867</td>\n",
       "      <td>334</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  0  \\\n",
       "0  ./phage_genomes/562_phage_genomes/AF063097.fasta   \n",
       "1  ./phage_genomes/562_phage_genomes/AF063097.fasta   \n",
       "2  ./phage_genomes/562_phage_genomes/AP019559.fasta   \n",
       "3  ./phage_genomes/562_phage_genomes/AP019559.fasta   \n",
       "4  ./phage_genomes/562_phage_genomes/AP019559.fasta   \n",
       "\n",
       "                                                   1         2    3    4  \n",
       "0   ./phage_genomes/562_phage_genomes/AF063097.fasta  100.0000  111  111  \n",
       "1  ./phage_genomes/562_phage_genomes/NC_001895.fasta  100.0000  111  111  \n",
       "2  ./phage_genomes/562_phage_genomes/NC_048627.fasta  100.0000  369  369  \n",
       "3   ./phage_genomes/562_phage_genomes/AP019559.fasta  100.0000  369  369  \n",
       "4   ./phage_genomes/562_phage_genomes/KJ190157.fasta   95.7867  334  369  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxon_id = '562'\n",
    "input_df = pd.read_csv('../Data/Other_possible_dbs/NCBI_phage_db/fastANI_results/{}_ANI.output'.format(taxon_id), sep='\\t', header=None)\n",
    "print('Shape:', input_df.shape)\n",
    "print('Non-null shape:', input_df[input_df[2].isnull()==False].shape)\n",
    "input_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4046, 5)\n",
      "(146959, 5)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_seq</th>\n",
       "      <th>ref_seq</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AF063097</td>\n",
       "      <td>NC_001895</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>111</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_048627</td>\n",
       "      <td>100.0000</td>\n",
       "      <td>369</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>KJ190157</td>\n",
       "      <td>95.7867</td>\n",
       "      <td>334</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_024139</td>\n",
       "      <td>95.7867</td>\n",
       "      <td>334</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>KY677846</td>\n",
       "      <td>95.3717</td>\n",
       "      <td>319</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_047822</td>\n",
       "      <td>95.3717</td>\n",
       "      <td>319</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_048818</td>\n",
       "      <td>95.2013</td>\n",
       "      <td>320</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MN316588</td>\n",
       "      <td>95.2013</td>\n",
       "      <td>320</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_047835</td>\n",
       "      <td>95.0970</td>\n",
       "      <td>326</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MF402939</td>\n",
       "      <td>95.0970</td>\n",
       "      <td>326</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MK373796</td>\n",
       "      <td>94.8456</td>\n",
       "      <td>319</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_048748</td>\n",
       "      <td>94.8456</td>\n",
       "      <td>319</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_017969</td>\n",
       "      <td>94.5494</td>\n",
       "      <td>324</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>HQ665011</td>\n",
       "      <td>94.5494</td>\n",
       "      <td>324</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_047754</td>\n",
       "      <td>94.5107</td>\n",
       "      <td>310</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>KR422353</td>\n",
       "      <td>94.5107</td>\n",
       "      <td>310</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MH745368</td>\n",
       "      <td>94.1427</td>\n",
       "      <td>315</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>NC_048731</td>\n",
       "      <td>94.1427</td>\n",
       "      <td>315</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MK373797</td>\n",
       "      <td>94.0003</td>\n",
       "      <td>310</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>AP019559</td>\n",
       "      <td>MK373792</td>\n",
       "      <td>93.9807</td>\n",
       "      <td>298</td>\n",
       "      <td>369</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   query_seq    ref_seq         2    3    4\n",
       "1   AF063097  NC_001895  100.0000  111  111\n",
       "2   AP019559  NC_048627  100.0000  369  369\n",
       "4   AP019559   KJ190157   95.7867  334  369\n",
       "5   AP019559  NC_024139   95.7867  334  369\n",
       "6   AP019559   KY677846   95.3717  319  369\n",
       "7   AP019559  NC_047822   95.3717  319  369\n",
       "8   AP019559  NC_048818   95.2013  320  369\n",
       "9   AP019559   MN316588   95.2013  320  369\n",
       "10  AP019559  NC_047835   95.0970  326  369\n",
       "11  AP019559   MF402939   95.0970  326  369\n",
       "12  AP019559   MK373796   94.8456  319  369\n",
       "13  AP019559  NC_048748   94.8456  319  369\n",
       "14  AP019559  NC_017969   94.5494  324  369\n",
       "15  AP019559   HQ665011   94.5494  324  369\n",
       "16  AP019559  NC_047754   94.5107  310  369\n",
       "17  AP019559   KR422353   94.5107  310  369\n",
       "18  AP019559   MH745368   94.1427  315  369\n",
       "19  AP019559  NC_048731   94.1427  315  369\n",
       "20  AP019559   MK373797   94.0003  310  369\n",
       "21  AP019559   MK373792   93.9807  298  369"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df['temp'] = input_df[0].str.split(\"/\").str[-1]\n",
    "input_df['query_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "input_df['temp'] = input_df[1].str.split(\"/\").str[-1]\n",
    "input_df['ref_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "input_df = input_df[['query_seq', 'ref_seq', 2, 3, 4]]\n",
    "print(input_df[input_df[2]==100.].shape)\n",
    "input_df = input_df[input_df['query_seq'] != input_df['ref_seq']]\n",
    "print(input_df.shape)\n",
    "input_df.head(n=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Re-create the cd-hit-est algorithm for clustering\n",
    "\n",
    "There's probably a faster way to do this but it seemed straightforward enough to quickly write up the greedy incremental clustering approach."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1772\n",
      "253 2070\n",
      "562\n",
      "271 1473\n",
      "28901\n",
      "132 526\n",
      "2055\n",
      "153 457\n",
      "1358\n",
      "117 441\n",
      "287\n",
      "80 429\n",
      "573\n",
      "137 376\n",
      "104336\n",
      "70 281\n",
      "1280\n",
      "57 262\n",
      "1771959\n",
      "49 241\n",
      "1308\n",
      "55 200\n",
      "1747\n",
      "61 183\n",
      "470\n",
      "49 152\n",
      "1313\n",
      "43 141\n",
      "1351\n",
      "43 128\n",
      "666\n",
      "20 121\n",
      "1428\n",
      "39 115\n",
      "1708203\n",
      "1 115\n",
      "1911\n",
      "34 110\n",
      "1314\n",
      "36 104\n",
      "599\n",
      "19 91\n",
      "1307\n",
      "47 91\n",
      "1665\n",
      "49 88\n",
      "552\n",
      "31 88\n",
      "1396\n",
      "42 82\n",
      "197\n",
      "16 77\n",
      "1464\n",
      "11 74\n",
      "317\n",
      "33 68\n",
      "1639\n",
      "14 66\n",
      "1423\n",
      "32 66\n",
      "996\n",
      "2 65\n",
      "670\n",
      "30 63\n",
      "623\n",
      "25 62\n",
      "644\n",
      "30 62\n",
      "1311\n",
      "26 60\n",
      "1334\n",
      "33 58\n",
      "305\n",
      "29 58\n",
      "29471\n",
      "6 55\n",
      "1833\n",
      "11 55\n",
      "36822\n",
      "22 53\n",
      "306\n",
      "19 53\n",
      "76594\n",
      "18 52\n"
     ]
    }
   ],
   "source": [
    "#Set sequence identity threshold\n",
    "thresh = 95\n",
    "\n",
    "for taxon_id in taxonomy_list[:]:\n",
    "    print(taxon_id)\n",
    "    ###Read in the ANI output\n",
    "    input_df = pd.read_csv('../Data/Other_possible_dbs/NCBI_phage_db/fastANI_results/{}_ANI.output'.format(taxon_id), sep='\\t', header=None)\n",
    "    ###Processing some columns\n",
    "    input_df['temp'] = input_df[0].str.split(\"/\").str[-1]\n",
    "    input_df['query_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "    input_df['temp'] = input_df[1].str.split(\"/\").str[-1]\n",
    "    input_df['ref_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "    input_df = input_df[['query_seq', 'ref_seq', 2, 3, 4]]\n",
    "    #Remove self vs self comparisons\n",
    "    input_df = input_df[input_df['query_seq'] != input_df['ref_seq']]\n",
    "    \n",
    "    ###Make a copy of the input\n",
    "    input_df_copy = input_df.copy(deep=True)\n",
    "    ###Select the species I care about from the full dataframe\n",
    "    species_df = full_df[full_df['Host_species_id']==taxon_id].sort_values('Length', ascending=False)\n",
    "    #Establish buckets\n",
    "    clusters = []\n",
    "    found = []\n",
    "    for index in species_df.index:\n",
    "        temp_accession = species_df.loc[index]['Accession']\n",
    "        ### Records should be sorted in order! (see above)\n",
    "        if temp_accession in found:\n",
    "            continue\n",
    "        ###Find all instances of this record in the ANI dataframe\n",
    "        temp_df = input_df_copy[(input_df_copy['query_seq']==temp_accession) | (input_df_copy['ref_seq']==temp_accession)]\n",
    "        ###Select all cases where the identity is greater than the specified threshold\n",
    "        temp_df = temp_df[temp_df[2] > thresh]\n",
    "        ###Add all these hits and call them a cluster!\n",
    "        temp_cluster = list(set(list(temp_df['query_seq'])+list(temp_df['ref_seq'])))\n",
    "        ###Now try to expand that cluster\n",
    "        stop = False\n",
    "        while stop == False:\n",
    "            starting = len(temp_cluster)\n",
    "            temp_df = input_df_copy[(input_df_copy['query_seq'].isin(temp_cluster)) | \n",
    "                                        (input_df_copy['ref_seq'].isin(temp_cluster))]\n",
    "            temp_df = temp_df[temp_df[2] > thresh]\n",
    "            temp_cluster = list(set(list(temp_df['query_seq'])+list(temp_df['ref_seq'])))\n",
    "            ###Break if I did not add anyone this iteration\n",
    "            if len(temp_cluster) == starting:\n",
    "                stop = True\n",
    "                \n",
    "        if len(temp_cluster) == 0:\n",
    "            temp_cluster = [temp_accession]\n",
    "            \n",
    "        clusters.append(temp_cluster)\n",
    "        found.extend(temp_cluster)\n",
    "        input_df_copy = input_df_copy[input_df_copy['query_seq'].isin(temp_cluster) == False]\n",
    "        input_df_copy = input_df_copy[input_df_copy['ref_seq'].isin(temp_cluster) == False]\n",
    "    print(len(clusters), len(found))\n",
    "    \n",
    "    ###Add to the dataframe\n",
    "    for i, cluster in enumerate(clusters):\n",
    "        temp_df = species_df[species_df['Accession'].isin(cluster)]\n",
    "        temp_df = temp_df.sort_values(['Sequence_Type', 'Length'], ascending=[False, False])\n",
    "        for j, index in enumerate(temp_df.index):\n",
    "            full_df.at[index, 'arbitrary_cluster_id'] = i + 1\n",
    "            full_df.at[index, 'ranking_in_cluster'] = j + 1\n",
    "\n",
    "    with open('../Data/Other_possible_dbs/NCBI_phage_db/fastANI_results/{}_clusters.json'.format(taxon_id), 'w') as outfile:\n",
    "        json.dump(clusters, outfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Write to file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "full_df['arbitrary_cluster_id'] = full_df['arbitrary_cluster_id'].astype(pd.Int64Dtype())\n",
    "full_df['ranking_in_cluster'] = full_df['ranking_in_cluster'].astype(pd.Int64Dtype())\n",
    "\n",
    "full_df = full_df[full_df['Host_species_id'].isin(taxonomy_list)]\n",
    "\n",
    "full_df.to_csv('../Data/Other_possible_dbs/NCBI_phage_db/'\n",
    "               'paper_dataset_11_2020_with_clusters.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualize what those results look like"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_df = pd.read_csv('../Data/{}_fastANI.output'.format(taxon_id), sep='\\t', header=None)\n",
    "# print('Shape:', input_df.shape)\n",
    "# print('Non-null shape:', input_df[input_df[2].isnull()==False].shape)\n",
    "# input_df['temp'] = input_df[0].str.split(\"/\").str[-1]\n",
    "# input_df['query_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "# input_df['temp'] = input_df[1].str.split(\"/\").str[-1]\n",
    "# input_df['ref_seq'] = input_df['temp'].str.split('.').str[0]\n",
    "# input_df = input_df[['query_seq', 'ref_seq', 2, 3, 4]]\n",
    "# print(input_df.shape)\n",
    "\n",
    "# sim_df = input_df.pivot(index='query_seq', columns='ref_seq', values=2)\n",
    "# sim_matrix = sim_df.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# temp = [str(item) for sublist in clusters for item in sublist if item in sim_df.index]\n",
    "# ordered_sim_matrix = sim_df.loc[temp][temp].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(len(sim_df.index), len(sim_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# fig, ax_arr = plt.subplots(nrows=1, ncols=2, figsize=(16,8))\n",
    "# ax_arr[0].matshow(sim_matrix)\n",
    "# ax_arr[0].set_title('Unsorted in any seemingly logical way')\n",
    "# ax_arr[1].matshow(ordered_sim_matrix)\n",
    "# ax_arr[1].set_title('After greedy algorithm')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ordered_sim_matrix[150, 190]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# input_df[2].hist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df = species_df[species_df['Accession'].isin(clusters[0])]\n",
    "temp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp_df.sort_values(['Sequence_Type', 'Length'], ascending=[False, False])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "base_numbering": 1,
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {},
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "threshold": 4,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
