{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook summary\n",
    "\n",
    "The purpose of this notebook is to create a set of `.tsv` files from various sources for both host and viral genomes. It is expected that the project organization follows the relative path where all initial files are located within `../Data`. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import gff3_parsing ###A separate .py library that should be available in the python path\n",
    "import pandas as pd\n",
    "import glob\n",
    "import subprocess\n",
    "\n",
    "import sys\n",
    "sys.path.append('../../iCUB/')\n",
    "import iCUB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Custom function definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_RBS_energy(df, energy_dict, col_name='aSD_binding',\\\n",
    "                   gaps=(4,10), expected_len=30, RBS_len=6):\n",
    "    '''\n",
    "    This function adds a ribosome binding site (RBS) energy column to the df based off of \n",
    "    pre-computed free energy vals (created with RNAcofold and stored in the corresponding \n",
    "    energy_dict. \n",
    "    \n",
    "    Inputs:\n",
    "        df - \n",
    "        energy_dict - \n",
    "        gaps - \n",
    "        expected_len - \n",
    "    \n",
    "    Outputs:\n",
    "        df - the transformed df object now containing the new column\n",
    "    '''\n",
    "    for index in df.index:\n",
    "        upstream = df.loc[index,'upstream_sequence']\n",
    "        test_string = upstream.replace('T', 'U')\n",
    "        ###Ensure that the sequence is the proper expected length\n",
    "        if len(test_string) != expected_len:\n",
    "            continue\n",
    "        ###Ensure that the sequence has no abnormal bases\n",
    "        if test_string.count('A') + test_string.count('U') +\\\n",
    "                                    test_string.count('C') + test_string.count('G') != expected_len:\n",
    "            continue\n",
    "            \n",
    "        ###Calculate the energy for the indicated gap offsets\n",
    "        energy_list = []\n",
    "        for gap in range(gaps[0],gaps[1]+1):\n",
    "             energy_list.append(energy_dict[test_string[-gap - RBS_len: -gap]])\n",
    "\n",
    "        df.at[index, col_name] = min(energy_list)        \n",
    "    return df\n",
    "\n",
    "\n",
    "def call_RNAfold(sequence):\n",
    "    sequence = sequence.replace('T', 'U')\n",
    "    MyOut = subprocess.Popen(['RNAfold', '-p', '--noPS', '--noDP', '--constraint'],\n",
    "            stdin=subprocess.PIPE,\n",
    "            stdout=subprocess.PIPE, \n",
    "            stderr=subprocess.STDOUT)\n",
    "    stdout, stderr = MyOut.communicate(input=str.encode(sequence))\n",
    "    return stdout\n",
    "\n",
    "def get_energy_RNAfold(stdout_string):\n",
    "    '''\n",
    "    '''\n",
    "    temp = stdout_string.decode('utf-8') \n",
    "    mfe_line = temp.split('\\n')[-5]\n",
    "    mfe_val = mfe_line[mfe_line.index(' '):]\n",
    "    mfe_val = mfe_val.strip().strip('()').strip()\n",
    "    #\n",
    "    ensemble_line = temp.split('\\n')[-4]\n",
    "    ensemble_val = ensemble_line[ensemble_line.index(' '):]\n",
    "    ensemble_val = ensemble_val.strip().strip('[]').strip()\n",
    "    return float(ensemble_val), float(mfe_val)    \n",
    "\n",
    "def add_secondary_structure(df):\n",
    "    '''\n",
    "    \n",
    "    '''\n",
    "    for index in df.index:\n",
    "        us_seq = df.loc[index]['upstream_sequence'] \n",
    "        cds_seq = df.loc[index]['coding_sequence']\n",
    "        if len(cds_seq) < 90:\n",
    "            continue\n",
    "        beg_cds_seq = cds_seq[:60]\n",
    "        if len(us_seq) == 30 and len(beg_cds_seq) == 60:\n",
    "            seq = us_seq + beg_cds_seq\n",
    "            rna_out = call_RNAfold(seq)\n",
    "            e1, e2 = get_energy_RNAfold(rna_out)\n",
    "            df.at[index, 'sec_struct'] = e1\n",
    "            #\n",
    "            rna_out = call_RNAfold(seq+'\\n'+ribo_bound_constraint)\n",
    "            e1, e2 = get_energy_RNAfold(rna_out)\n",
    "            df.at[index, 'sec_struct_bound'] = e1\n",
    "\n",
    "    return df\n",
    "\n",
    "def add_iCUB_and_GC(df):\n",
    "    '''\n",
    "    '''\n",
    "    for index in df.index:\n",
    "        cds_seq = df.loc[index]['coding_sequence']\n",
    "        if len(cds_seq) == 0:\n",
    "            continue\n",
    "        if len(cds_seq) != cds_seq.count('A') + cds_seq.count('T') + cds_seq.count('C') + cds_seq.count('G'):\n",
    "            continue\n",
    "        if len(cds_seq)%3 != 0:\n",
    "            continue\n",
    "        df.at[index, 'iCUB'] = iCUB.iCUB_Calculator(cds_seq).get_iCUB()\n",
    "        #\n",
    "        df.at[index, 'GC_cds'] = (cds_seq.count('G') + cds_seq.count('C')) / len(cds_seq)\n",
    "        #\n",
    "        us_seq = df.loc[index]['upstream_sequence']\n",
    "        if len(us_seq) == 30:\n",
    "            df.at[index, 'GC_upstream'] = (us_seq.count('G') + us_seq.count('C')) / len(us_seq)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook-wide parameters "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "sep = '\\t'\n",
    "upstream_len = 30\n",
    "\n",
    "with open('../Data/energy_files/energyRef_CCUCCU_ensemble_noneConstraint.json', 'r') as infile:\n",
    "       energy_dict = json.load(infile)\n",
    "        \n",
    "base_viral_genome_dir = '../Data/MVP_data/host_linked_genomes/'\n",
    "host_genome_dir = '../Data/MVP_data/host_genomes/'\n",
    "\n",
    "\n",
    "\n",
    "# ribo_bound_constraint = ('.'*16) + ('x'*28)+ ('.'*16)\n",
    "# assert len(ribo_bound_constraint)==60\n",
    "\n",
    "# ribo_bound_constraint = ('.'*17) + ('x'*25)+ ('.'*58)\n",
    "# assert len(ribo_bound_constraint)==100\n",
    "\n",
    "ribo_bound_constraint = ('.'*16) + ('x'*28)+ ('.'*46)\n",
    "assert len(ribo_bound_constraint)==90"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating `.tsv` files for host genomes\n",
    "\n",
    "**Adding relevant columns for ribosome binding sites, secondary structure, codon usage bias, and GC content. This whole process takes 5-10 minutes per genome. Could definitely be optimized but for now not a bottleneck, just anticipate a few hours for this cell to run.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "host_ids = [36809,\\\n",
    "           717959,\\\n",
    "           305,\\\n",
    "           1590,\\\n",
    "           435591,\\\n",
    "           90371,\\\n",
    "           1314,\\\n",
    "           357276,\\\n",
    "           657318,\\\n",
    "           1639,\\\n",
    "           1428,\\\n",
    "           470,\\\n",
    "           573,\\\n",
    "           1280,\\\n",
    "           287,\\\n",
    "           562]\n",
    "\n",
    "for host_id in host_ids:\n",
    "    print(host_id)\n",
    "    ###Creates the dataframe based off a gff3 and fasta file\n",
    "    host_df, host_genome = gff3_parsing.compile_sequences([host_genome_dir + '{}.gff3'.format(host_id)],\\\n",
    "                                                        [host_genome_dir + '{}.fasta'.format(host_id)],\\\n",
    "                                                          upstream_len)\n",
    "    ###Adds the ribosome binding site energy column\n",
    "    host_df = add_RBS_energy(host_df, energy_dict, col_name='aSD_binding', gaps=(4,10))\n",
    "\n",
    "    ###Adds the secondary structure column\n",
    "    host_df = add_secondary_structure(host_df)\n",
    "    \n",
    "    ###Add codon usage bias column\n",
    "    host_df = add_iCUB_and_GC(host_df)\n",
    "    \n",
    "    ###Writes to a file\n",
    "    host_df.to_csv(host_genome_dir + '{}.tsv'.format(host_id), sep=sep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Treat the 2 chromosome hosts slightly differently to ensure no errors occur\n",
    "\n",
    "(There is currently only one of these. And I don't love the solution because it's basically entirely copy/paste but here we are.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "host_ids_2chrom = [28450]\n",
    "\n",
    "for host_id in host_ids_2chrom:\n",
    "    print(host_id)\n",
    "    gffs = [host_genome_dir + '{}.1.gff3'.format(host_id),\\\n",
    "            host_genome_dir + '{}.2.gff3'.format(host_id)]\n",
    "\n",
    "    fastas = [host_genome_dir + '{}.1.fasta'.format(host_id),\\\n",
    "              host_genome_dir + '{}.2.fasta'.format(host_id)]\n",
    "\n",
    "    host_df, host_genome = gff3_parsing.compile_sequences(gffs, fastas, upstream_len)\n",
    "    \n",
    "    ###Adds the ribosome binding site energy column\n",
    "    host_df = add_RBS_energy(host_df, energy_dict, col_name='aSD_binding', gaps=(4,10))\n",
    "\n",
    "    ###Adds the secondary structure column\n",
    "    host_df = add_secondary_structure(host_df)\n",
    "    \n",
    "    ###Add codon usage bias column\n",
    "    host_df = add_iCUB_and_GC(host_df)\n",
    "    \n",
    "    ###Writes to a file\n",
    "    host_df.to_csv(host_genome_dir + '{}.tsv'.format(host_id), sep=sep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Creating `.tsv` files for viral genomes\n",
    "**There are a lot more of these viral genomes but they're super small so run-time for this cell should also be in the few hour range at most. And it only needs to happen once so this isn't a bottleneck and I'm not concerning myself with speed.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Concatenate the two existing host lists\n",
    "host_ids = host_ids + host_ids_2chrom\n",
    "for host_id in host_ids:\n",
    "    print('#####', host_id)\n",
    "    for gff_file in glob.glob(base_viral_genome_dir+'{}_rep_viruses/*.gff3'.format(host_id)):\n",
    "        print(gff_file)\n",
    "        fasta_file = gff_file.replace('.gff3', '.fasta')\n",
    "        tsv_file = gff_file.replace('.gff3', '.tsv')\n",
    "        \n",
    "        viral_df, viral_genome = gff3_parsing.compile_sequences([gff_file], [fasta_file], upstream_len)\n",
    "        \n",
    "        ###Adds the ribosome binding site energy column\n",
    "        viral_df = add_RBS_energy(viral_df, energy_dict, col_name='aSD_binding', gaps=(4,10))\n",
    "\n",
    "        ###Adds the secondary structure column\n",
    "        viral_df = add_secondary_structure(viral_df)\n",
    "\n",
    "        ###Add codon usage bias column\n",
    "        viral_df = add_iCUB_and_GC(viral_df)\n",
    "\n",
    "        ###Writes to a file\n",
    "        viral_df.to_csv(tsv_file, sep=sep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make \"clean\" `.tsv` files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def common_cleaning(df):\n",
    "    df = df.reset_index(drop=True)\n",
    "    df = df[df['upstream_sequence'].isnull()==False]\n",
    "    df = df[df['coding_sequence'].isnull()==False]\n",
    "    df = df[df['iCUB'].isnull()==False]\n",
    "    df = df[df['GC_cds'].isnull()==False]\n",
    "    df = df[df['GC_upstream'].isnull()==False]\n",
    "    df = df[df['aSD_binding'].isnull()==False]\n",
    "    df = df[df['sec_struct'].isnull()==False]\n",
    "    return df\n",
    "\n",
    "def clean_host_tsv(df):\n",
    "    \"\"\"\n",
    "    WRITE A BRIEF PURPOSE/SUMMARY\n",
    "    \n",
    "    Development notes: I could also think about testing for stop codons within coding sequences\n",
    "    and filtering accordingly. Also a way to test for possible non-standard genetic code usage\n",
    "    \n",
    "    Input/s:\n",
    "        df - a pandas dataframe with numeric indices, outputted from Compile_data.ipynb and read\n",
    "                back in\n",
    "        \n",
    "    Output/s:\n",
    "        df - a clean version of the dataframe with one new column (iCUB) and (potentially)\n",
    "                several rows removed\n",
    "    \"\"\"\n",
    "    ###Run the main cleaning/additions\n",
    "    df = common_cleaning(df)\n",
    "    \n",
    "    ###Filter out possible prophage genes by removing anything involving the word phage\n",
    "    ###Numerous possibilities/ways to do this and not all genomes might have any decent\n",
    "    ###descriptions in the qualifiers.\n",
    "    filter_word = 'phage'\n",
    "    df = df[(df['qualifiers'].str.contains(filter_word)==False)]\n",
    "    \n",
    "    ###Now ensure that each locus tag is only used once and when in doubt remove them both\n",
    "    df['locus_tag'] = df['qualifiers'].str.split('locus_tag=', n=1, expand=True)[1]\\\n",
    "                            .str.split(';', n=1, expand=True)[0]\n",
    "    df = df.drop_duplicates(subset = ['locus_tag'], keep = False)\n",
    "    return df\n",
    "\n",
    "def clean_virus_tsv(df):\n",
    "    \"\"\"\n",
    "    This is the same basic structure as the \"clean_host_tsv\" function. \n",
    "    \n",
    "    Input/s:\n",
    "        df - a pandas dataframe with numeric indices, outputted from Compile_data.ipynb and read\n",
    "                back in\n",
    "        \n",
    "    Output/s:\n",
    "        df - a clean version of the dataframe with one new column (iCUB) and (potentially)\n",
    "                several rows removed\n",
    "                \n",
    "    \"\"\"  \n",
    "    ###Run the main cleaning/additions\n",
    "    df = common_cleaning(df)\n",
    "\n",
    "    ###Now ensure that each viral_id tag is only used once (and when in doubt remove them BOTH)\n",
    "    df['viral_id'] = df['qualifiers'].str.split('ID=', n=1, expand=True)[1]\\\n",
    "                            .str.split(';', n=1, expand=True)[0]\n",
    "    \n",
    "    df = df.drop_duplicates(subset = [\"viral_id\"], keep = False)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Side analysis just to make sure that all these genomes use the standard translation table**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for host_tsv_file in glob.glob(host_genome_dir + '*.tsv')[:]:\n",
    "    if '.clean.' in host_tsv_file:\n",
    "        continue\n",
    "    print(host_tsv_file)\n",
    "    df = pd.read_csv(host_tsv_file, sep = '\\t', index_col = 0)\n",
    "    df['transl_table'] = df['qualifiers'].str.split('transl_table=', n=1, expand=True)[1]\\\n",
    "                        .str.split(';', n=1, expand=True)[0]\n",
    "    print(df.shape[0], list(df['transl_table'].value_counts().items()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Now clean the host data tables**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for host_tsv_file in glob.glob(host_genome_dir + '*.tsv'):\n",
    "    if '.clean.' in host_tsv_file:\n",
    "        continue\n",
    "    print(host_tsv_file)\n",
    "    ###\n",
    "    df = pd.read_csv(host_tsv_file, sep='\\t', index_col=0)\n",
    "    initial_shape = df.shape\n",
    "    df = clean_host_tsv(df)\n",
    "    final_shape = df.shape\n",
    "    ###\n",
    "    gene_ratio = final_shape[0]/initial_shape[0]\n",
    "    print(gene_ratio)\n",
    "    if gene_ratio <= 0.8: #This basically shouldn't happen\n",
    "        break\n",
    "    ###\n",
    "    clean_tsv_loc = host_tsv_file.replace('.tsv', '.clean.tsv')\n",
    "    df.to_csv(clean_tsv_loc, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And the phages**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for virus_folder in glob.glob(base_viral_genome_dir + '*_rep_viruses/'):\n",
    "    print('###', virus_folder)\n",
    "    for virus_tsv_file in glob.glob(virus_folder + '*.tsv'):\n",
    "        if '.clean.' in virus_tsv_file:\n",
    "            continue\n",
    "        if '.fasta.' in virus_tsv_file:\n",
    "            continue\n",
    "        print(virus_tsv_file)\n",
    "        ###\n",
    "        df = pd.read_csv(virus_tsv_file, sep='\\t', index_col=0)\n",
    "        initial_shape = df.shape\n",
    "        df = clean_virus_tsv(df)\n",
    "        final_shape = df.shape\n",
    "        ###\n",
    "        gene_ratio = final_shape[0]/initial_shape[0]\n",
    "        if gene_ratio <= 0.8:\n",
    "            print('Strange case', gene_ratio)\n",
    "            break\n",
    "        ###\n",
    "        clean_tsv_loc = virus_tsv_file.replace('.tsv', '.clean.tsv')\n",
    "        df.to_csv(clean_tsv_loc, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Add in the well-annotated reference genome/s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Data/benchmark_viruses/T7.gff3\n"
     ]
    }
   ],
   "source": [
    "for gff_file in glob.glob('../Data/benchmark_viruses/*.gff3'):\n",
    "    print(gff_file)\n",
    "    fasta_file = gff_file.replace('.gff3', '.fasta')\n",
    "    tsv_file = gff_file.replace('.gff3', '.tsv')\n",
    "\n",
    "    viral_df, viral_genome = gff3_parsing.compile_sequences([gff_file], [fasta_file], upstream_len)\n",
    "\n",
    "    ###Adds the ribosome binding site energy column\n",
    "    viral_df = add_RBS_energy(viral_df, energy_dict, col_name='aSD_binding', gaps=(4,10))\n",
    "\n",
    "    ###Adds the secondary structure column\n",
    "    viral_df = add_secondary_structure(viral_df)\n",
    "\n",
    "    ###Add codon usage bias column\n",
    "    viral_df = add_iCUB_and_GC(viral_df)\n",
    "\n",
    "    ###Writes to a file\n",
    "    viral_df.to_csv(tsv_file, sep=sep)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Data/benchmark_viruses/T7.tsv\n"
     ]
    }
   ],
   "source": [
    "for virus_tsv_file in glob.glob('../Data/benchmark_viruses/*.tsv'):\n",
    "        if '.clean.' in virus_tsv_file:\n",
    "            continue\n",
    "        if '.fasta.' in virus_tsv_file:\n",
    "            continue\n",
    "        print(virus_tsv_file)\n",
    "        ###\n",
    "        df = pd.read_csv(virus_tsv_file, sep='\\t', index_col=0)\n",
    "        initial_shape = df.shape\n",
    "        df = clean_virus_tsv(df)\n",
    "        final_shape = df.shape\n",
    "        ###\n",
    "        gene_ratio = final_shape[0]/initial_shape[0]\n",
    "        if gene_ratio <= 0.8:\n",
    "            print('Strange case', gene_ratio)\n",
    "            break\n",
    "        ###\n",
    "        clean_tsv_loc = virus_tsv_file.replace('.tsv', '.clean.tsv')\n",
    "        df.to_csv(clean_tsv_loc, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_id</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "      <th>idk</th>\n",
       "      <th>strand</th>\n",
       "      <th>trash</th>\n",
       "      <th>qualifiers</th>\n",
       "      <th>coding_sequence</th>\n",
       "      <th>upstream_sequence</th>\n",
       "      <th>aSD_binding</th>\n",
       "      <th>sec_struct</th>\n",
       "      <th>sec_struct_bound</th>\n",
       "      <th>iCUB</th>\n",
       "      <th>GC_cds</th>\n",
       "      <th>GC_upstream</th>\n",
       "      <th>viral_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>925</td>\n",
       "      <td>1278</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041954.1;Parent=rna-T7p01;Dbxref=Uni...</td>\n",
       "      <td>ATGGCTATGTCTAACATGACTTACAACAACGTTTTCGACCACGCTT...</td>\n",
       "      <td>TTCACTAATAACTGCACGAGGTAACACAAG</td>\n",
       "      <td>-5.42</td>\n",
       "      <td>-10.13</td>\n",
       "      <td>-3.81</td>\n",
       "      <td>50.525906</td>\n",
       "      <td>0.468927</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041954.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1278</td>\n",
       "      <td>1433</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041955.1;Parent=gene-T7p02;Dbxref=Un...</td>\n",
       "      <td>ATGTCTACTACCAACGTGCAATACGGTCTGACCGCTCAAACTGTAC...</td>\n",
       "      <td>AGTCGAGGAGTACGAGGAGGATGAAGAGTA</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-19.01</td>\n",
       "      <td>-12.96</td>\n",
       "      <td>43.066882</td>\n",
       "      <td>0.455128</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>cds-NP_041955.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2021</td>\n",
       "      <td>3100</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041959.1;Parent=rna-T7p03;Dbxref=GOA...</td>\n",
       "      <td>ATGAACATTACCGACATCATGAACGCTATCGACGCAATCAAAGCAC...</td>\n",
       "      <td>TCTCATAACGAACATAAAGGACACAATGCA</td>\n",
       "      <td>-4.00</td>\n",
       "      <td>-7.46</td>\n",
       "      <td>-3.80</td>\n",
       "      <td>55.711182</td>\n",
       "      <td>0.493519</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>cds-NP_041959.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1496</td>\n",
       "      <td>1639</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041956.1;Parent=gene-T7p04;Dbxref=Un...</td>\n",
       "      <td>ATGTATATGCTTACTATCGGTCTACTCACCGCTCTAGGTCTAGCTG...</td>\n",
       "      <td>TATGATTATCACTTTACTTATGAGGGAGTA</td>\n",
       "      <td>-5.82</td>\n",
       "      <td>-19.87</td>\n",
       "      <td>-8.50</td>\n",
       "      <td>39.155014</td>\n",
       "      <td>0.458333</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>cds-NP_041956.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1636</td>\n",
       "      <td>1797</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041958.1;Parent=gene-T7p06;Dbxref=Un...</td>\n",
       "      <td>ATGATGAAGCACTACGTTATGCCAATCCACACGTCCAACGGGGCAA...</td>\n",
       "      <td>CATAGGAATCATCAAAGGGGCACTACGCAA</td>\n",
       "      <td>-5.53</td>\n",
       "      <td>-17.68</td>\n",
       "      <td>-7.52</td>\n",
       "      <td>42.406008</td>\n",
       "      <td>0.481481</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_041958.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>3171</td>\n",
       "      <td>5822</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041960.1;Parent=rna-T7p07;Dbxref=GOA...</td>\n",
       "      <td>ATGAACACGATTAACATCGCTAAGAACGACTTCTCTGACATCGAAC...</td>\n",
       "      <td>GTACGATTTACTAACTGGAAGAGGCACTAA</td>\n",
       "      <td>-5.13</td>\n",
       "      <td>-10.80</td>\n",
       "      <td>-5.39</td>\n",
       "      <td>53.909493</td>\n",
       "      <td>0.514706</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041960.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6137</td>\n",
       "      <td>6394</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041962.1;Parent=rna-T7p08;Dbxref=GOA...</td>\n",
       "      <td>ATGGGACGTTTATATAGTGGTAATCTGGCAGCATTCAAGGCAGCAA...</td>\n",
       "      <td>CTCACAAGCGTAGCTGGGAGGGTCAGTAAG</td>\n",
       "      <td>-9.51</td>\n",
       "      <td>-17.09</td>\n",
       "      <td>-9.78</td>\n",
       "      <td>45.974315</td>\n",
       "      <td>0.422481</td>\n",
       "      <td>0.566667</td>\n",
       "      <td>cds-NP_041962.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6007</td>\n",
       "      <td>6135</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041961.1;Parent=gene-T7p09;Dbxref=Un...</td>\n",
       "      <td>ATGCGTAACTTCGAAAAGATGACCAAACGTTCTAACCGTAATGCTC...</td>\n",
       "      <td>TTAAAGAATTACTAAGAGAGGACTTTAAGT</td>\n",
       "      <td>-5.83</td>\n",
       "      <td>-15.83</td>\n",
       "      <td>-6.80</td>\n",
       "      <td>37.097063</td>\n",
       "      <td>0.480620</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>cds-NP_041961.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>6475</td>\n",
       "      <td>7554</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041963.1;Parent=rna-T7p10;Dbxref=GOA...</td>\n",
       "      <td>ATGATGAACATTAAGACTAACCCGTTTAAAGCCGTGTCTTTCGTAG...</td>\n",
       "      <td>TAGTCATTTAACCAATAGGAGATAAACATT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-10.28</td>\n",
       "      <td>-5.72</td>\n",
       "      <td>55.811993</td>\n",
       "      <td>0.469444</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>cds-NP_041963.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>7608</td>\n",
       "      <td>7763</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041964.1;Parent=gene-T7p11;Dbxref=Un...</td>\n",
       "      <td>ATGTTTAAGAAGGTTGGTAAATTCCTTGCGGCTTTGGCAGCTATCC...</td>\n",
       "      <td>GGCCTTTCTGCGTTTATAAGGAGACACTTT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-18.92</td>\n",
       "      <td>-9.77</td>\n",
       "      <td>50.382798</td>\n",
       "      <td>0.455128</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041964.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>7791</td>\n",
       "      <td>7880</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041965.1;Parent=gene-T7p12;Dbxref=Un...</td>\n",
       "      <td>ATGATGTACTTAATGCCATTACTCATCGTCATTGTAGGATGCCTTG...</td>\n",
       "      <td>TAATACGACTCACTAAAGGAGGTACACACC</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-15.78</td>\n",
       "      <td>-6.42</td>\n",
       "      <td>36.766447</td>\n",
       "      <td>0.444444</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041965.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>7906</td>\n",
       "      <td>8166</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041966.1;Parent=gene-T7p13;Dbxref=Un...</td>\n",
       "      <td>ATGTTTCGACTTCATTACAACAAAAGCGTTAAGAATTTCACGGTTC...</td>\n",
       "      <td>CTTAATACGACTCACTAAAGGAGACACTAT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-15.17</td>\n",
       "      <td>-9.57</td>\n",
       "      <td>48.840195</td>\n",
       "      <td>0.482759</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>cds-NP_041966.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8166</td>\n",
       "      <td>8756</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041967.1;Parent=gene-T7p14;Dbxref=Un...</td>\n",
       "      <td>ATGGGACTGTTAGATGGTGAAGCCTGGGAAAAAGAAAACCCGCCAG...</td>\n",
       "      <td>GTTACTCAAACGAATCAAGGAGGTGTTCTG</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-16.76</td>\n",
       "      <td>-10.50</td>\n",
       "      <td>55.478469</td>\n",
       "      <td>0.441624</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041967.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8749</td>\n",
       "      <td>8895</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041968.1;Parent=gene-T7p15;Dbxref=Un...</td>\n",
       "      <td>ATGCATAACTTCAAGTCAACCCCACCTGCCGACAGCCTATCTGATG...</td>\n",
       "      <td>AACTCTTTGAGAAACATAAGGATAAATGTT</td>\n",
       "      <td>-4.08</td>\n",
       "      <td>-10.44</td>\n",
       "      <td>-4.79</td>\n",
       "      <td>43.274884</td>\n",
       "      <td>0.469388</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>cds-NP_041968.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>8898</td>\n",
       "      <td>9092</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041969.1;Parent=gene-T7p16;Dbxref=Un...</td>\n",
       "      <td>ATGTCAAACGTAAATACAGGTTCACTTAGTGTGGACAATAAGAAGT...</td>\n",
       "      <td>TGAACTTTGGAAATCGAGAGGTCAATGACT</td>\n",
       "      <td>-5.33</td>\n",
       "      <td>-17.35</td>\n",
       "      <td>-10.25</td>\n",
       "      <td>45.695741</td>\n",
       "      <td>0.487179</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041969.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9158</td>\n",
       "      <td>9856</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041970.1;Parent=gene-T7p17;Dbxref=GO...</td>\n",
       "      <td>ATGGCTAAGAAGATTTTCACCTCTGCGCTGGGTACCGCTGAACCTT...</td>\n",
       "      <td>ACCAAACGAAACCTAAAGGAGATTAACATT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-11.94</td>\n",
       "      <td>-7.99</td>\n",
       "      <td>49.136668</td>\n",
       "      <td>0.503577</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_041970.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>9857</td>\n",
       "      <td>10276</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041971.1;Parent=gene-T7p18;Dbxref=GO...</td>\n",
       "      <td>GTGGAACTGCGGGAGAAAATCCTTGAGCGAATCAAGGTGACTTCCT...</td>\n",
       "      <td>GAGGAAGCAGACGAAGACGGAGACTTCTAA</td>\n",
       "      <td>-6.61</td>\n",
       "      <td>-22.67</td>\n",
       "      <td>-13.91</td>\n",
       "      <td>52.481875</td>\n",
       "      <td>0.490476</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>cds-NP_041971.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10257</td>\n",
       "      <td>10706</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041972.1;Parent=gene-T7p19;Dbxref=GO...</td>\n",
       "      <td>ATGGCAGGTTACGGCGCTAAAGGAATCCGAAAGGTTGGAGCGTTTC...</td>\n",
       "      <td>CTATTTGTGATATACGCAAAGGGAGGCGAC</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-27.57</td>\n",
       "      <td>-16.69</td>\n",
       "      <td>54.008683</td>\n",
       "      <td>0.431111</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_041972.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>10706</td>\n",
       "      <td>11161</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041973.1;Parent=gene-T7p20;Dbxref=GO...</td>\n",
       "      <td>ATGGCTCGTGTACAGTTTAAACAACGTGAATCTACTGACGCAATCT...</td>\n",
       "      <td>TAGATTAAAAAGGAAAGGAGGAAAGAAATA</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-13.67</td>\n",
       "      <td>-4.97</td>\n",
       "      <td>50.637820</td>\n",
       "      <td>0.521930</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>cds-NP_041973.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>11225</td>\n",
       "      <td>11590</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041974.1;Parent=gene-T7p21;Dbxref=GO...</td>\n",
       "      <td>ATGCGCAAGTCTTATAAACAATTCTATAAGGCTCCGAGGAGGCATA...</td>\n",
       "      <td>TTTCCCTTTGTTCGCATTGGAGGTCAAATA</td>\n",
       "      <td>-8.81</td>\n",
       "      <td>-23.52</td>\n",
       "      <td>-10.06</td>\n",
       "      <td>54.679623</td>\n",
       "      <td>0.456284</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041974.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>11565</td>\n",
       "      <td>13265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041975.1;Parent=gene-T7p22;Dbxref=GO...</td>\n",
       "      <td>ATGGACAATTCGCACGATTCCGATAGTGTATTTCTTTACCACATTC...</td>\n",
       "      <td>CGTGGATTTATAGAACTAGGAGGGAATTGC</td>\n",
       "      <td>-9.51</td>\n",
       "      <td>-18.36</td>\n",
       "      <td>-6.47</td>\n",
       "      <td>56.287505</td>\n",
       "      <td>0.488536</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041975.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>11635</td>\n",
       "      <td>11757</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041976.1;Parent=gene-T7p23;Dbxref=Un...</td>\n",
       "      <td>ATGGGAACTCGCTGTTCTCTGACGGACACACGTTCTGCTACGTATG...</td>\n",
       "      <td>ACATTCCTTGTGACAACTGTGGGAGTAGTG</td>\n",
       "      <td>-6.51</td>\n",
       "      <td>-25.16</td>\n",
       "      <td>-10.33</td>\n",
       "      <td>38.086833</td>\n",
       "      <td>0.487805</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_041976.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>11754</td>\n",
       "      <td>13265</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041977.1;Parent=gene-T7p24;Dbxref=GO...</td>\n",
       "      <td>ATGACTTACAACGTGTGGAACTTCGGGGAATCCAATGGACGCTACT...</td>\n",
       "      <td>TCAAAACGGAAACCCTCAGGAGGTAAACCA</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-21.77</td>\n",
       "      <td>-12.32</td>\n",
       "      <td>55.534298</td>\n",
       "      <td>0.490741</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_041977.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>12988</td>\n",
       "      <td>13326</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041978.1;Parent=gene-T7p25;Dbxref=Un...</td>\n",
       "      <td>ATGAGGAAGGTCGCCCCGTTTCTATTACTGACCTACGTGGTTCTGG...</td>\n",
       "      <td>ACCTTAAGAACCCAGACAAAGGTAAAGCAC</td>\n",
       "      <td>-2.52</td>\n",
       "      <td>-21.90</td>\n",
       "      <td>-11.67</td>\n",
       "      <td>54.358884</td>\n",
       "      <td>0.477876</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041978.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13352</td>\n",
       "      <td>13564</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041979.1;Parent=gene-T7p26;Dbxref=Un...</td>\n",
       "      <td>ATGTTCAAACTGATTAAGAAGTTAGGCCAACTGCTGGTTCGTATGT...</td>\n",
       "      <td>TCTAATACGACTCACTAAAGGAGACACACC</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-14.97</td>\n",
       "      <td>-8.09</td>\n",
       "      <td>46.132562</td>\n",
       "      <td>0.497653</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041979.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13584</td>\n",
       "      <td>13853</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041980.1;Parent=gene-T7p27;Dbxref=Un...</td>\n",
       "      <td>ATGTCTAACGTAGCTGAAACTATCCGTCTATCCGATACAGCTGACC...</td>\n",
       "      <td>TCTTTGAGTAATCAAACAGGAGAAACCATT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-15.36</td>\n",
       "      <td>-7.01</td>\n",
       "      <td>50.425452</td>\n",
       "      <td>0.485185</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_041980.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>13927</td>\n",
       "      <td>14334</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041981.1;Parent=gene-T7p28;Dbxref=Un...</td>\n",
       "      <td>ATGCGTGACCCTAAAGTTATCCAAGCAGAAATCGCTAAACTGGAAG...</td>\n",
       "      <td>ACTATTCGACTCACTATAGGAGATATTACC</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-15.53</td>\n",
       "      <td>-8.16</td>\n",
       "      <td>46.775101</td>\n",
       "      <td>0.455882</td>\n",
       "      <td>0.366667</td>\n",
       "      <td>cds-NP_041981.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>14353</td>\n",
       "      <td>16467</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041982.1;Parent=gene-T7p29;Dbxref=GO...</td>\n",
       "      <td>ATGATCGTTTCTGACATCGAAGCTAACGCCCTCTTAGAGAGCGTCA...</td>\n",
       "      <td>AAGTCACGATAATCAATAGGAGAAATCAAT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-16.27</td>\n",
       "      <td>-9.09</td>\n",
       "      <td>55.203032</td>\n",
       "      <td>0.505910</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>cds-NP_041982.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>16483</td>\n",
       "      <td>16839</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041983.1;Parent=gene-T7p30;Dbxref=Un...</td>\n",
       "      <td>ATGAACGAAAGACACTTAACAGGTGCTGCTTCTGAAATGCTAGTAG...</td>\n",
       "      <td>GCGATTTGCCACTGATACAGGAGGCTACTC</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-20.85</td>\n",
       "      <td>-9.21</td>\n",
       "      <td>52.963385</td>\n",
       "      <td>0.425770</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>cds-NP_041983.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>16851</td>\n",
       "      <td>17150</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041985.1;Parent=gene-T7p32;Dbxref=GO...</td>\n",
       "      <td>ATGGCTATGACAAAGAAATTTAAAGTGTCCTTCGACGTTACCGCAA...</td>\n",
       "      <td>CAAAGGCATAAAACTATAGGAGAAATTATT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-14.83</td>\n",
       "      <td>-9.94</td>\n",
       "      <td>50.230438</td>\n",
       "      <td>0.443333</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>cds-NP_041985.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17150</td>\n",
       "      <td>17359</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041986.1;Parent=gene-T7p33;Dbxref=GO...</td>\n",
       "      <td>ATGTCTGACTACCTGAAAGTGCTGCAAGCAATCAAAAGTTGCCCTA...</td>\n",
       "      <td>ATCTCCTGCAACAGTACGGGAGGTGTTCTG</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-23.31</td>\n",
       "      <td>-11.07</td>\n",
       "      <td>46.284931</td>\n",
       "      <td>0.514286</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>cds-NP_041986.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17359</td>\n",
       "      <td>17517</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041987.1;Parent=gene-T7p34;Dbxref=Un...</td>\n",
       "      <td>ATGTCTCGTGACCTTGTGACTATTCCACGCGATGTGTGGAACGATA...</td>\n",
       "      <td>CTTTCTGAAACGAATGGGAGGATGTGTCTA</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-25.56</td>\n",
       "      <td>-12.23</td>\n",
       "      <td>42.758423</td>\n",
       "      <td>0.452830</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041987.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>17504</td>\n",
       "      <td>18406</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041988.1;Parent=gene-T7p35;Dbxref=GO...</td>\n",
       "      <td>ATGGCACTTCTTGACCTTAAACAATTCTATGAGTTACGTGAAGGCT...</td>\n",
       "      <td>AATACGTAGCGGAACTAGAGGAGAAACTTA</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-15.63</td>\n",
       "      <td>-7.99</td>\n",
       "      <td>52.640637</td>\n",
       "      <td>0.495017</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041988.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18394</td>\n",
       "      <td>18507</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041989.1;Parent=gene-T7p36;Dbxref=Un...</td>\n",
       "      <td>GTGGAGACCGTAGCGTATATTGGTCTGGGTCTTTGTGTTCTCGGAG...</td>\n",
       "      <td>GTACAACTTTATTGACAAGGAGATTTACCT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-22.77</td>\n",
       "      <td>-9.92</td>\n",
       "      <td>43.669331</td>\n",
       "      <td>0.491228</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_041989.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18605</td>\n",
       "      <td>18859</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041990.1;Parent=gene-T7p37;Dbxref=Un...</td>\n",
       "      <td>ATGTTAACACCTATTAACCAATTACTTAAGAACCCTAACGATATTC...</td>\n",
       "      <td>TAAGATTTAACTCTAAGAGGAATCTTTATT</td>\n",
       "      <td>-5.83</td>\n",
       "      <td>-9.50</td>\n",
       "      <td>-2.14</td>\n",
       "      <td>50.568240</td>\n",
       "      <td>0.443137</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>cds-NP_041990.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>18864</td>\n",
       "      <td>19130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041991.1;Parent=gene-T7p38;Dbxref=Un...</td>\n",
       "      <td>ATGTGTTTCTCACCGAAAATTAAAACTCCGAAGATGGATACCAATC...</td>\n",
       "      <td>AACTAAGAGATGATGGGGAGGATTGACACT</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-17.19</td>\n",
       "      <td>-6.62</td>\n",
       "      <td>50.521668</td>\n",
       "      <td>0.513109</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041991.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>19130</td>\n",
       "      <td>19531</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041992.1;Parent=gene-T7p39;Dbxref=Un...</td>\n",
       "      <td>ATGTCTGAGTTCACATGTGTGGAGGCTAAGAGTCGCTTCCGTGCAA...</td>\n",
       "      <td>CCGTAAGTCCGCATTTGGAGGTAAGAAGTG</td>\n",
       "      <td>-8.81</td>\n",
       "      <td>-33.02</td>\n",
       "      <td>-18.45</td>\n",
       "      <td>51.947299</td>\n",
       "      <td>0.487562</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>cds-NP_041992.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>19535</td>\n",
       "      <td>19834</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041993.1;Parent=gene-T7p40;Dbxref=Un...</td>\n",
       "      <td>ATGGGTAAGAAAGTTAAGAAGGCCGTGAAGAAAGTCACCAAGTCCG...</td>\n",
       "      <td>ATGCGTGTATACTTTAAGGAGGTATAAGTT</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-12.85</td>\n",
       "      <td>-5.18</td>\n",
       "      <td>49.735661</td>\n",
       "      <td>0.513333</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_041993.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>19848</td>\n",
       "      <td>20240</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041994.1;Parent=gene-T7p41;Dbxref=GO...</td>\n",
       "      <td>GTGGAAGACTGCATTGAATGGACCGGAGGTGTCAACTCTAAGGGTT...</td>\n",
       "      <td>GCGGTATCAACATTTAATCAGGAGGTTATC</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-19.56</td>\n",
       "      <td>-9.92</td>\n",
       "      <td>52.197060</td>\n",
       "      <td>0.447837</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041994.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>20240</td>\n",
       "      <td>21850</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041995.1;Parent=gene-T7p42;Dbxref=GO...</td>\n",
       "      <td>ATGGCTGAGAAACGAACAGGACTTGCGGAGGATGGCGCAAAGTCTG...</td>\n",
       "      <td>ACTACAGCGTAAGACATGGAGACACATTTA</td>\n",
       "      <td>-6.41</td>\n",
       "      <td>-23.01</td>\n",
       "      <td>-13.09</td>\n",
       "      <td>56.023748</td>\n",
       "      <td>0.492241</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_041995.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>21950</td>\n",
       "      <td>22873</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041996.1;Parent=gene-T7p43;Dbxref=GO...</td>\n",
       "      <td>ATGGCTGAATCTAATGCAGACGTATATGCATCTTTTGGCGTGAACT...</td>\n",
       "      <td>CTGTAGTTCAACTTTAAGGAGACAATAATA</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-15.83</td>\n",
       "      <td>-11.16</td>\n",
       "      <td>52.222377</td>\n",
       "      <td>0.515152</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>cds-NP_041996.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>22967</td>\n",
       "      <td>24004</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041998.1;Parent=gene-T7p45;Dbxref=GO...</td>\n",
       "      <td>ATGGCTAGCATGACTGGTGGACAGCAAATGGGTACTAACCAAGGTA...</td>\n",
       "      <td>TTTGTTTAACTTTAAGAAGGAGATATACAT</td>\n",
       "      <td>-7.00</td>\n",
       "      <td>-14.62</td>\n",
       "      <td>-9.62</td>\n",
       "      <td>48.443366</td>\n",
       "      <td>0.530829</td>\n",
       "      <td>0.233333</td>\n",
       "      <td>cds-NP_041998.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>24228</td>\n",
       "      <td>24818</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_041999.1;Parent=gene-T7p46;Dbxref=Un...</td>\n",
       "      <td>ATGCGCTCATACGATATGAACGTTGAGACTGCCGCTGAGTTATCAG...</td>\n",
       "      <td>GAGGGGTTTTTTGCTGAAAGGAGGAACTAT</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-20.40</td>\n",
       "      <td>-12.36</td>\n",
       "      <td>51.781196</td>\n",
       "      <td>0.483926</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_041999.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>24842</td>\n",
       "      <td>27226</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042000.1;Parent=gene-T7p47;Dbxref=Un...</td>\n",
       "      <td>ATGGCACTCATTAGCCAATCAATCAAGAACTTGAAGGGTGGTATCA...</td>\n",
       "      <td>TCGCTAACATTAATAAATAAGGAGGCTCTA</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-18.94</td>\n",
       "      <td>-8.75</td>\n",
       "      <td>56.887108</td>\n",
       "      <td>0.470440</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_042000.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27307</td>\n",
       "      <td>27723</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042001.1;Parent=gene-T7p48;Dbxref=Un...</td>\n",
       "      <td>ATGATGACTATAAGACCTACTAAAAGTACAGACTTTGAGGTATTCA...</td>\n",
       "      <td>AGAACAATACGACTACGGGAGGGTTTTCTT</td>\n",
       "      <td>-9.51</td>\n",
       "      <td>-16.16</td>\n",
       "      <td>-6.21</td>\n",
       "      <td>51.281910</td>\n",
       "      <td>0.458034</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_042001.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>27728</td>\n",
       "      <td>28318</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042002.1;Parent=gene-T7p49;Dbxref=Un...</td>\n",
       "      <td>ATGTGTTGGGCAGCCGCAATACCTATCGCTATATCTGGCGCTCAGG...</td>\n",
       "      <td>TATTTACAATCACGAAAGGAGGATAACCAT</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-19.45</td>\n",
       "      <td>-9.83</td>\n",
       "      <td>54.614836</td>\n",
       "      <td>0.505922</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_042002.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>28325</td>\n",
       "      <td>30568</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042003.1;Parent=gene-T7p50;Dbxref=Un...</td>\n",
       "      <td>ATGAGTAAAATTGAATCTGCCCTTCAAGCGGCACAACCGGGACTCT...</td>\n",
       "      <td>AAAGGAACCAAGACGGGGAGGTAATGAGCT</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-23.94</td>\n",
       "      <td>-17.09</td>\n",
       "      <td>55.540588</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>cds-NP_042003.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>30595</td>\n",
       "      <td>34551</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042004.1;Parent=gene-T7p51;Dbxref=GO...</td>\n",
       "      <td>ATGGATAAGTACGATAAGAACGTACCAAGTGATTATGATGGTCTGT...</td>\n",
       "      <td>GTAACTAAAGGCTACATAAGGAGGCCCTAA</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-17.45</td>\n",
       "      <td>-8.80</td>\n",
       "      <td>56.602989</td>\n",
       "      <td>0.495577</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_042004.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>34624</td>\n",
       "      <td>36285</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042005.1;Parent=gene-T7p52;Dbxref=Un...</td>\n",
       "      <td>ATGGCTAACGTAATTAAAACCGTTTTGACTTACCAGTTAGATGGCT...</td>\n",
       "      <td>AGTCTCTTAGATTTACTTTAAGGAGGTCAA</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-17.39</td>\n",
       "      <td>-8.54</td>\n",
       "      <td>57.179167</td>\n",
       "      <td>0.480746</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>cds-NP_042005.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>36344</td>\n",
       "      <td>36547</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042006.1;Parent=gene-T7p53;Dbxref=GO...</td>\n",
       "      <td>GTGCTATCATTAGACTTTAACAACGAATTGATTAAGGCTGCTCCAA...</td>\n",
       "      <td>TCCACGGATGGACTCTCAAGGAGGTACAAG</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-18.63</td>\n",
       "      <td>-12.15</td>\n",
       "      <td>48.008637</td>\n",
       "      <td>0.441176</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>cds-NP_042006.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>36553</td>\n",
       "      <td>36822</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042007.1;Parent=gene-T7p54;Dbxref=GO...</td>\n",
       "      <td>ATGGAAAAGGATAAGAGCCTTATTACATTCTTAGAGATGTTGGACA...</td>\n",
       "      <td>CTGGAAGAAAGCCAATAAGGAGTGATATGT</td>\n",
       "      <td>-6.51</td>\n",
       "      <td>-18.31</td>\n",
       "      <td>-11.27</td>\n",
       "      <td>46.598805</td>\n",
       "      <td>0.451852</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_042007.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>36917</td>\n",
       "      <td>37348</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042008.1;Parent=gene-T7p55;Dbxref=GO...</td>\n",
       "      <td>ATGCTAGAATTTTTACGTAAGCTAATCCCTTGGGTTCTCGCTGGGA...</td>\n",
       "      <td>CGTGAAATCTGAAAGTTAACGGGAGGCATT</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-21.34</td>\n",
       "      <td>-12.81</td>\n",
       "      <td>54.753265</td>\n",
       "      <td>0.488426</td>\n",
       "      <td>0.433333</td>\n",
       "      <td>cds-NP_042008.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>37032</td>\n",
       "      <td>37283</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042009.1;Parent=gene-T7p56;Dbxref=Un...</td>\n",
       "      <td>ATGAGTACGTTAAGAGAGTTGAGGCTGCGAAGAGCACTCAAAGAGC...</td>\n",
       "      <td>TGGACGCTAAATGGAAACAGGAGGTACACA</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-17.91</td>\n",
       "      <td>-11.49</td>\n",
       "      <td>49.879406</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_042009.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>37370</td>\n",
       "      <td>39130</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042010.1;Parent=gene-T7p57;Dbxref=GO...</td>\n",
       "      <td>GTGTCTACTCAATCCAATCGTAATGCGCTCGTAGTGGCGCAACTGA...</td>\n",
       "      <td>CGTAAGTAGGAAATCAAGTAAGGAGGCAAT</td>\n",
       "      <td>-8.83</td>\n",
       "      <td>-19.09</td>\n",
       "      <td>-10.86</td>\n",
       "      <td>54.254624</td>\n",
       "      <td>0.498580</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_042010.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>38016</td>\n",
       "      <td>38273</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042011.1;Parent=gene-T7p58;Dbxref=Un...</td>\n",
       "      <td>GTGGGTACACAACCATTATCTGGCCTGCTCTGTACCCAAGGACACG...</td>\n",
       "      <td>TGACTCTCTATAAGGAACTTGAGGATAACC</td>\n",
       "      <td>-5.66</td>\n",
       "      <td>-20.49</td>\n",
       "      <td>-7.01</td>\n",
       "      <td>48.602199</td>\n",
       "      <td>0.527132</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>cds-NP_042011.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>38553</td>\n",
       "      <td>38726</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042012.1;Parent=gene-T7p59;Dbxref=Un...</td>\n",
       "      <td>ATGGCTACTCCGATAAGACCCTTGAGTTACTCGCTAAGAAGGCAAA...</td>\n",
       "      <td>TCTACCTTATGGAAGCTGGAGGTTTCCGTG</td>\n",
       "      <td>-8.81</td>\n",
       "      <td>-24.12</td>\n",
       "      <td>-10.84</td>\n",
       "      <td>47.883227</td>\n",
       "      <td>0.482759</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>cds-NP_042012.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>NC_001604.1</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CDS</td>\n",
       "      <td>39389</td>\n",
       "      <td>39538</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>ID=cds-NP_042013.1;Parent=gene-T7p60;Dbxref=Un...</td>\n",
       "      <td>ATGTTCCGCTTATTGTTGAACCTACTGCGGCATAGAGTCACCTACC...</td>\n",
       "      <td>CTTAAGGACCAACATAAAGGGAGGAGACTC</td>\n",
       "      <td>-9.31</td>\n",
       "      <td>-27.24</td>\n",
       "      <td>-12.79</td>\n",
       "      <td>51.309820</td>\n",
       "      <td>0.473333</td>\n",
       "      <td>0.466667</td>\n",
       "      <td>cds-NP_042013.1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      genome_id  source type  start   stop idk strand  trash  \\\n",
       "0   NC_001604.1  RefSeq  CDS    925   1278   .      +      0   \n",
       "1   NC_001604.1  RefSeq  CDS   1278   1433   .      +      0   \n",
       "2   NC_001604.1  RefSeq  CDS   2021   3100   .      +      0   \n",
       "3   NC_001604.1  RefSeq  CDS   1496   1639   .      +      0   \n",
       "6   NC_001604.1  RefSeq  CDS   1636   1797   .      +      0   \n",
       "7   NC_001604.1  RefSeq  CDS   3171   5822   .      +      0   \n",
       "8   NC_001604.1  RefSeq  CDS   6137   6394   .      +      0   \n",
       "9   NC_001604.1  RefSeq  CDS   6007   6135   .      +      0   \n",
       "10  NC_001604.1  RefSeq  CDS   6475   7554   .      +      0   \n",
       "11  NC_001604.1  RefSeq  CDS   7608   7763   .      +      0   \n",
       "12  NC_001604.1  RefSeq  CDS   7791   7880   .      +      0   \n",
       "13  NC_001604.1  RefSeq  CDS   7906   8166   .      +      0   \n",
       "14  NC_001604.1  RefSeq  CDS   8166   8756   .      +      0   \n",
       "15  NC_001604.1  RefSeq  CDS   8749   8895   .      +      0   \n",
       "16  NC_001604.1  RefSeq  CDS   8898   9092   .      +      0   \n",
       "17  NC_001604.1  RefSeq  CDS   9158   9856   .      +      0   \n",
       "18  NC_001604.1  RefSeq  CDS   9857  10276   .      +      0   \n",
       "19  NC_001604.1  RefSeq  CDS  10257  10706   .      +      0   \n",
       "20  NC_001604.1  RefSeq  CDS  10706  11161   .      +      0   \n",
       "21  NC_001604.1  RefSeq  CDS  11225  11590   .      +      0   \n",
       "22  NC_001604.1  RefSeq  CDS  11565  13265   .      +      0   \n",
       "23  NC_001604.1  RefSeq  CDS  11635  11757   .      +      0   \n",
       "24  NC_001604.1  RefSeq  CDS  11754  13265   .      +      0   \n",
       "25  NC_001604.1  RefSeq  CDS  12988  13326   .      +      0   \n",
       "26  NC_001604.1  RefSeq  CDS  13352  13564   .      +      0   \n",
       "27  NC_001604.1  RefSeq  CDS  13584  13853   .      +      0   \n",
       "28  NC_001604.1  RefSeq  CDS  13927  14334   .      +      0   \n",
       "29  NC_001604.1  RefSeq  CDS  14353  16467   .      +      0   \n",
       "30  NC_001604.1  RefSeq  CDS  16483  16839   .      +      0   \n",
       "33  NC_001604.1  RefSeq  CDS  16851  17150   .      +      0   \n",
       "34  NC_001604.1  RefSeq  CDS  17150  17359   .      +      0   \n",
       "35  NC_001604.1  RefSeq  CDS  17359  17517   .      +      0   \n",
       "36  NC_001604.1  RefSeq  CDS  17504  18406   .      +      0   \n",
       "37  NC_001604.1  RefSeq  CDS  18394  18507   .      +      0   \n",
       "38  NC_001604.1  RefSeq  CDS  18605  18859   .      +      0   \n",
       "39  NC_001604.1  RefSeq  CDS  18864  19130   .      +      0   \n",
       "40  NC_001604.1  RefSeq  CDS  19130  19531   .      +      0   \n",
       "41  NC_001604.1  RefSeq  CDS  19535  19834   .      +      0   \n",
       "42  NC_001604.1  RefSeq  CDS  19848  20240   .      +      0   \n",
       "43  NC_001604.1  RefSeq  CDS  20240  21850   .      +      0   \n",
       "44  NC_001604.1  RefSeq  CDS  21950  22873   .      +      0   \n",
       "47  NC_001604.1  RefSeq  CDS  22967  24004   .      +      0   \n",
       "48  NC_001604.1  RefSeq  CDS  24228  24818   .      +      0   \n",
       "49  NC_001604.1  RefSeq  CDS  24842  27226   .      +      0   \n",
       "50  NC_001604.1  RefSeq  CDS  27307  27723   .      +      0   \n",
       "51  NC_001604.1  RefSeq  CDS  27728  28318   .      +      0   \n",
       "52  NC_001604.1  RefSeq  CDS  28325  30568   .      +      0   \n",
       "53  NC_001604.1  RefSeq  CDS  30595  34551   .      +      0   \n",
       "54  NC_001604.1  RefSeq  CDS  34624  36285   .      +      0   \n",
       "55  NC_001604.1  RefSeq  CDS  36344  36547   .      +      0   \n",
       "56  NC_001604.1  RefSeq  CDS  36553  36822   .      +      0   \n",
       "57  NC_001604.1  RefSeq  CDS  36917  37348   .      +      0   \n",
       "58  NC_001604.1  RefSeq  CDS  37032  37283   .      +      0   \n",
       "59  NC_001604.1  RefSeq  CDS  37370  39130   .      +      0   \n",
       "60  NC_001604.1  RefSeq  CDS  38016  38273   .      +      0   \n",
       "61  NC_001604.1  RefSeq  CDS  38553  38726   .      +      0   \n",
       "62  NC_001604.1  RefSeq  CDS  39389  39538   .      +      0   \n",
       "\n",
       "                                           qualifiers  \\\n",
       "0   ID=cds-NP_041954.1;Parent=rna-T7p01;Dbxref=Uni...   \n",
       "1   ID=cds-NP_041955.1;Parent=gene-T7p02;Dbxref=Un...   \n",
       "2   ID=cds-NP_041959.1;Parent=rna-T7p03;Dbxref=GOA...   \n",
       "3   ID=cds-NP_041956.1;Parent=gene-T7p04;Dbxref=Un...   \n",
       "6   ID=cds-NP_041958.1;Parent=gene-T7p06;Dbxref=Un...   \n",
       "7   ID=cds-NP_041960.1;Parent=rna-T7p07;Dbxref=GOA...   \n",
       "8   ID=cds-NP_041962.1;Parent=rna-T7p08;Dbxref=GOA...   \n",
       "9   ID=cds-NP_041961.1;Parent=gene-T7p09;Dbxref=Un...   \n",
       "10  ID=cds-NP_041963.1;Parent=rna-T7p10;Dbxref=GOA...   \n",
       "11  ID=cds-NP_041964.1;Parent=gene-T7p11;Dbxref=Un...   \n",
       "12  ID=cds-NP_041965.1;Parent=gene-T7p12;Dbxref=Un...   \n",
       "13  ID=cds-NP_041966.1;Parent=gene-T7p13;Dbxref=Un...   \n",
       "14  ID=cds-NP_041967.1;Parent=gene-T7p14;Dbxref=Un...   \n",
       "15  ID=cds-NP_041968.1;Parent=gene-T7p15;Dbxref=Un...   \n",
       "16  ID=cds-NP_041969.1;Parent=gene-T7p16;Dbxref=Un...   \n",
       "17  ID=cds-NP_041970.1;Parent=gene-T7p17;Dbxref=GO...   \n",
       "18  ID=cds-NP_041971.1;Parent=gene-T7p18;Dbxref=GO...   \n",
       "19  ID=cds-NP_041972.1;Parent=gene-T7p19;Dbxref=GO...   \n",
       "20  ID=cds-NP_041973.1;Parent=gene-T7p20;Dbxref=GO...   \n",
       "21  ID=cds-NP_041974.1;Parent=gene-T7p21;Dbxref=GO...   \n",
       "22  ID=cds-NP_041975.1;Parent=gene-T7p22;Dbxref=GO...   \n",
       "23  ID=cds-NP_041976.1;Parent=gene-T7p23;Dbxref=Un...   \n",
       "24  ID=cds-NP_041977.1;Parent=gene-T7p24;Dbxref=GO...   \n",
       "25  ID=cds-NP_041978.1;Parent=gene-T7p25;Dbxref=Un...   \n",
       "26  ID=cds-NP_041979.1;Parent=gene-T7p26;Dbxref=Un...   \n",
       "27  ID=cds-NP_041980.1;Parent=gene-T7p27;Dbxref=Un...   \n",
       "28  ID=cds-NP_041981.1;Parent=gene-T7p28;Dbxref=Un...   \n",
       "29  ID=cds-NP_041982.1;Parent=gene-T7p29;Dbxref=GO...   \n",
       "30  ID=cds-NP_041983.1;Parent=gene-T7p30;Dbxref=Un...   \n",
       "33  ID=cds-NP_041985.1;Parent=gene-T7p32;Dbxref=GO...   \n",
       "34  ID=cds-NP_041986.1;Parent=gene-T7p33;Dbxref=GO...   \n",
       "35  ID=cds-NP_041987.1;Parent=gene-T7p34;Dbxref=Un...   \n",
       "36  ID=cds-NP_041988.1;Parent=gene-T7p35;Dbxref=GO...   \n",
       "37  ID=cds-NP_041989.1;Parent=gene-T7p36;Dbxref=Un...   \n",
       "38  ID=cds-NP_041990.1;Parent=gene-T7p37;Dbxref=Un...   \n",
       "39  ID=cds-NP_041991.1;Parent=gene-T7p38;Dbxref=Un...   \n",
       "40  ID=cds-NP_041992.1;Parent=gene-T7p39;Dbxref=Un...   \n",
       "41  ID=cds-NP_041993.1;Parent=gene-T7p40;Dbxref=Un...   \n",
       "42  ID=cds-NP_041994.1;Parent=gene-T7p41;Dbxref=GO...   \n",
       "43  ID=cds-NP_041995.1;Parent=gene-T7p42;Dbxref=GO...   \n",
       "44  ID=cds-NP_041996.1;Parent=gene-T7p43;Dbxref=GO...   \n",
       "47  ID=cds-NP_041998.1;Parent=gene-T7p45;Dbxref=GO...   \n",
       "48  ID=cds-NP_041999.1;Parent=gene-T7p46;Dbxref=Un...   \n",
       "49  ID=cds-NP_042000.1;Parent=gene-T7p47;Dbxref=Un...   \n",
       "50  ID=cds-NP_042001.1;Parent=gene-T7p48;Dbxref=Un...   \n",
       "51  ID=cds-NP_042002.1;Parent=gene-T7p49;Dbxref=Un...   \n",
       "52  ID=cds-NP_042003.1;Parent=gene-T7p50;Dbxref=Un...   \n",
       "53  ID=cds-NP_042004.1;Parent=gene-T7p51;Dbxref=GO...   \n",
       "54  ID=cds-NP_042005.1;Parent=gene-T7p52;Dbxref=Un...   \n",
       "55  ID=cds-NP_042006.1;Parent=gene-T7p53;Dbxref=GO...   \n",
       "56  ID=cds-NP_042007.1;Parent=gene-T7p54;Dbxref=GO...   \n",
       "57  ID=cds-NP_042008.1;Parent=gene-T7p55;Dbxref=GO...   \n",
       "58  ID=cds-NP_042009.1;Parent=gene-T7p56;Dbxref=Un...   \n",
       "59  ID=cds-NP_042010.1;Parent=gene-T7p57;Dbxref=GO...   \n",
       "60  ID=cds-NP_042011.1;Parent=gene-T7p58;Dbxref=Un...   \n",
       "61  ID=cds-NP_042012.1;Parent=gene-T7p59;Dbxref=Un...   \n",
       "62  ID=cds-NP_042013.1;Parent=gene-T7p60;Dbxref=Un...   \n",
       "\n",
       "                                      coding_sequence  \\\n",
       "0   ATGGCTATGTCTAACATGACTTACAACAACGTTTTCGACCACGCTT...   \n",
       "1   ATGTCTACTACCAACGTGCAATACGGTCTGACCGCTCAAACTGTAC...   \n",
       "2   ATGAACATTACCGACATCATGAACGCTATCGACGCAATCAAAGCAC...   \n",
       "3   ATGTATATGCTTACTATCGGTCTACTCACCGCTCTAGGTCTAGCTG...   \n",
       "6   ATGATGAAGCACTACGTTATGCCAATCCACACGTCCAACGGGGCAA...   \n",
       "7   ATGAACACGATTAACATCGCTAAGAACGACTTCTCTGACATCGAAC...   \n",
       "8   ATGGGACGTTTATATAGTGGTAATCTGGCAGCATTCAAGGCAGCAA...   \n",
       "9   ATGCGTAACTTCGAAAAGATGACCAAACGTTCTAACCGTAATGCTC...   \n",
       "10  ATGATGAACATTAAGACTAACCCGTTTAAAGCCGTGTCTTTCGTAG...   \n",
       "11  ATGTTTAAGAAGGTTGGTAAATTCCTTGCGGCTTTGGCAGCTATCC...   \n",
       "12  ATGATGTACTTAATGCCATTACTCATCGTCATTGTAGGATGCCTTG...   \n",
       "13  ATGTTTCGACTTCATTACAACAAAAGCGTTAAGAATTTCACGGTTC...   \n",
       "14  ATGGGACTGTTAGATGGTGAAGCCTGGGAAAAAGAAAACCCGCCAG...   \n",
       "15  ATGCATAACTTCAAGTCAACCCCACCTGCCGACAGCCTATCTGATG...   \n",
       "16  ATGTCAAACGTAAATACAGGTTCACTTAGTGTGGACAATAAGAAGT...   \n",
       "17  ATGGCTAAGAAGATTTTCACCTCTGCGCTGGGTACCGCTGAACCTT...   \n",
       "18  GTGGAACTGCGGGAGAAAATCCTTGAGCGAATCAAGGTGACTTCCT...   \n",
       "19  ATGGCAGGTTACGGCGCTAAAGGAATCCGAAAGGTTGGAGCGTTTC...   \n",
       "20  ATGGCTCGTGTACAGTTTAAACAACGTGAATCTACTGACGCAATCT...   \n",
       "21  ATGCGCAAGTCTTATAAACAATTCTATAAGGCTCCGAGGAGGCATA...   \n",
       "22  ATGGACAATTCGCACGATTCCGATAGTGTATTTCTTTACCACATTC...   \n",
       "23  ATGGGAACTCGCTGTTCTCTGACGGACACACGTTCTGCTACGTATG...   \n",
       "24  ATGACTTACAACGTGTGGAACTTCGGGGAATCCAATGGACGCTACT...   \n",
       "25  ATGAGGAAGGTCGCCCCGTTTCTATTACTGACCTACGTGGTTCTGG...   \n",
       "26  ATGTTCAAACTGATTAAGAAGTTAGGCCAACTGCTGGTTCGTATGT...   \n",
       "27  ATGTCTAACGTAGCTGAAACTATCCGTCTATCCGATACAGCTGACC...   \n",
       "28  ATGCGTGACCCTAAAGTTATCCAAGCAGAAATCGCTAAACTGGAAG...   \n",
       "29  ATGATCGTTTCTGACATCGAAGCTAACGCCCTCTTAGAGAGCGTCA...   \n",
       "30  ATGAACGAAAGACACTTAACAGGTGCTGCTTCTGAAATGCTAGTAG...   \n",
       "33  ATGGCTATGACAAAGAAATTTAAAGTGTCCTTCGACGTTACCGCAA...   \n",
       "34  ATGTCTGACTACCTGAAAGTGCTGCAAGCAATCAAAAGTTGCCCTA...   \n",
       "35  ATGTCTCGTGACCTTGTGACTATTCCACGCGATGTGTGGAACGATA...   \n",
       "36  ATGGCACTTCTTGACCTTAAACAATTCTATGAGTTACGTGAAGGCT...   \n",
       "37  GTGGAGACCGTAGCGTATATTGGTCTGGGTCTTTGTGTTCTCGGAG...   \n",
       "38  ATGTTAACACCTATTAACCAATTACTTAAGAACCCTAACGATATTC...   \n",
       "39  ATGTGTTTCTCACCGAAAATTAAAACTCCGAAGATGGATACCAATC...   \n",
       "40  ATGTCTGAGTTCACATGTGTGGAGGCTAAGAGTCGCTTCCGTGCAA...   \n",
       "41  ATGGGTAAGAAAGTTAAGAAGGCCGTGAAGAAAGTCACCAAGTCCG...   \n",
       "42  GTGGAAGACTGCATTGAATGGACCGGAGGTGTCAACTCTAAGGGTT...   \n",
       "43  ATGGCTGAGAAACGAACAGGACTTGCGGAGGATGGCGCAAAGTCTG...   \n",
       "44  ATGGCTGAATCTAATGCAGACGTATATGCATCTTTTGGCGTGAACT...   \n",
       "47  ATGGCTAGCATGACTGGTGGACAGCAAATGGGTACTAACCAAGGTA...   \n",
       "48  ATGCGCTCATACGATATGAACGTTGAGACTGCCGCTGAGTTATCAG...   \n",
       "49  ATGGCACTCATTAGCCAATCAATCAAGAACTTGAAGGGTGGTATCA...   \n",
       "50  ATGATGACTATAAGACCTACTAAAAGTACAGACTTTGAGGTATTCA...   \n",
       "51  ATGTGTTGGGCAGCCGCAATACCTATCGCTATATCTGGCGCTCAGG...   \n",
       "52  ATGAGTAAAATTGAATCTGCCCTTCAAGCGGCACAACCGGGACTCT...   \n",
       "53  ATGGATAAGTACGATAAGAACGTACCAAGTGATTATGATGGTCTGT...   \n",
       "54  ATGGCTAACGTAATTAAAACCGTTTTGACTTACCAGTTAGATGGCT...   \n",
       "55  GTGCTATCATTAGACTTTAACAACGAATTGATTAAGGCTGCTCCAA...   \n",
       "56  ATGGAAAAGGATAAGAGCCTTATTACATTCTTAGAGATGTTGGACA...   \n",
       "57  ATGCTAGAATTTTTACGTAAGCTAATCCCTTGGGTTCTCGCTGGGA...   \n",
       "58  ATGAGTACGTTAAGAGAGTTGAGGCTGCGAAGAGCACTCAAAGAGC...   \n",
       "59  GTGTCTACTCAATCCAATCGTAATGCGCTCGTAGTGGCGCAACTGA...   \n",
       "60  GTGGGTACACAACCATTATCTGGCCTGCTCTGTACCCAAGGACACG...   \n",
       "61  ATGGCTACTCCGATAAGACCCTTGAGTTACTCGCTAAGAAGGCAAA...   \n",
       "62  ATGTTCCGCTTATTGTTGAACCTACTGCGGCATAGAGTCACCTACC...   \n",
       "\n",
       "                 upstream_sequence  aSD_binding  sec_struct  sec_struct_bound  \\\n",
       "0   TTCACTAATAACTGCACGAGGTAACACAAG        -5.42      -10.13             -3.81   \n",
       "1   AGTCGAGGAGTACGAGGAGGATGAAGAGTA        -9.31      -19.01            -12.96   \n",
       "2   TCTCATAACGAACATAAAGGACACAATGCA        -4.00       -7.46             -3.80   \n",
       "3   TATGATTATCACTTTACTTATGAGGGAGTA        -5.82      -19.87             -8.50   \n",
       "6   CATAGGAATCATCAAAGGGGCACTACGCAA        -5.53      -17.68             -7.52   \n",
       "7   GTACGATTTACTAACTGGAAGAGGCACTAA        -5.13      -10.80             -5.39   \n",
       "8   CTCACAAGCGTAGCTGGGAGGGTCAGTAAG        -9.51      -17.09             -9.78   \n",
       "9   TTAAAGAATTACTAAGAGAGGACTTTAAGT        -5.83      -15.83             -6.80   \n",
       "10  TAGTCATTTAACCAATAGGAGATAAACATT        -7.00      -10.28             -5.72   \n",
       "11  GGCCTTTCTGCGTTTATAAGGAGACACTTT        -7.00      -18.92             -9.77   \n",
       "12  TAATACGACTCACTAAAGGAGGTACACACC        -8.83      -15.78             -6.42   \n",
       "13  CTTAATACGACTCACTAAAGGAGACACTAT        -7.00      -15.17             -9.57   \n",
       "14  GTTACTCAAACGAATCAAGGAGGTGTTCTG        -8.83      -16.76            -10.50   \n",
       "15  AACTCTTTGAGAAACATAAGGATAAATGTT        -4.08      -10.44             -4.79   \n",
       "16  TGAACTTTGGAAATCGAGAGGTCAATGACT        -5.33      -17.35            -10.25   \n",
       "17  ACCAAACGAAACCTAAAGGAGATTAACATT        -7.00      -11.94             -7.99   \n",
       "18  GAGGAAGCAGACGAAGACGGAGACTTCTAA        -6.61      -22.67            -13.91   \n",
       "19  CTATTTGTGATATACGCAAAGGGAGGCGAC        -8.83      -27.57            -16.69   \n",
       "20  TAGATTAAAAAGGAAAGGAGGAAAGAAATA        -9.31      -13.67             -4.97   \n",
       "21  TTTCCCTTTGTTCGCATTGGAGGTCAAATA        -8.81      -23.52            -10.06   \n",
       "22  CGTGGATTTATAGAACTAGGAGGGAATTGC        -9.51      -18.36             -6.47   \n",
       "23  ACATTCCTTGTGACAACTGTGGGAGTAGTG        -6.51      -25.16            -10.33   \n",
       "24  TCAAAACGGAAACCCTCAGGAGGTAAACCA        -8.83      -21.77            -12.32   \n",
       "25  ACCTTAAGAACCCAGACAAAGGTAAAGCAC        -2.52      -21.90            -11.67   \n",
       "26  TCTAATACGACTCACTAAAGGAGACACACC        -7.00      -14.97             -8.09   \n",
       "27  TCTTTGAGTAATCAAACAGGAGAAACCATT        -7.00      -15.36             -7.01   \n",
       "28  ACTATTCGACTCACTATAGGAGATATTACC        -7.00      -15.53             -8.16   \n",
       "29  AAGTCACGATAATCAATAGGAGAAATCAAT        -7.00      -16.27             -9.09   \n",
       "30  GCGATTTGCCACTGATACAGGAGGCTACTC        -8.83      -20.85             -9.21   \n",
       "33  CAAAGGCATAAAACTATAGGAGAAATTATT        -7.00      -14.83             -9.94   \n",
       "34  ATCTCCTGCAACAGTACGGGAGGTGTTCTG        -8.83      -23.31            -11.07   \n",
       "35  CTTTCTGAAACGAATGGGAGGATGTGTCTA        -9.31      -25.56            -12.23   \n",
       "36  AATACGTAGCGGAACTAGAGGAGAAACTTA        -7.00      -15.63             -7.99   \n",
       "37  GTACAACTTTATTGACAAGGAGATTTACCT        -7.00      -22.77             -9.92   \n",
       "38  TAAGATTTAACTCTAAGAGGAATCTTTATT        -5.83       -9.50             -2.14   \n",
       "39  AACTAAGAGATGATGGGGAGGATTGACACT        -9.31      -17.19             -6.62   \n",
       "40  CCGTAAGTCCGCATTTGGAGGTAAGAAGTG        -8.81      -33.02            -18.45   \n",
       "41  ATGCGTGTATACTTTAAGGAGGTATAAGTT        -8.83      -12.85             -5.18   \n",
       "42  GCGGTATCAACATTTAATCAGGAGGTTATC        -8.83      -19.56             -9.92   \n",
       "43  ACTACAGCGTAAGACATGGAGACACATTTA        -6.41      -23.01            -13.09   \n",
       "44  CTGTAGTTCAACTTTAAGGAGACAATAATA        -7.00      -15.83            -11.16   \n",
       "47  TTTGTTTAACTTTAAGAAGGAGATATACAT        -7.00      -14.62             -9.62   \n",
       "48  GAGGGGTTTTTTGCTGAAAGGAGGAACTAT        -9.31      -20.40            -12.36   \n",
       "49  TCGCTAACATTAATAAATAAGGAGGCTCTA        -8.83      -18.94             -8.75   \n",
       "50  AGAACAATACGACTACGGGAGGGTTTTCTT        -9.51      -16.16             -6.21   \n",
       "51  TATTTACAATCACGAAAGGAGGATAACCAT        -9.31      -19.45             -9.83   \n",
       "52  AAAGGAACCAAGACGGGGAGGTAATGAGCT        -8.83      -23.94            -17.09   \n",
       "53  GTAACTAAAGGCTACATAAGGAGGCCCTAA        -8.83      -17.45             -8.80   \n",
       "54  AGTCTCTTAGATTTACTTTAAGGAGGTCAA        -8.83      -17.39             -8.54   \n",
       "55  TCCACGGATGGACTCTCAAGGAGGTACAAG        -8.83      -18.63            -12.15   \n",
       "56  CTGGAAGAAAGCCAATAAGGAGTGATATGT        -6.51      -18.31            -11.27   \n",
       "57  CGTGAAATCTGAAAGTTAACGGGAGGCATT        -8.83      -21.34            -12.81   \n",
       "58  TGGACGCTAAATGGAAACAGGAGGTACACA        -8.83      -17.91            -11.49   \n",
       "59  CGTAAGTAGGAAATCAAGTAAGGAGGCAAT        -8.83      -19.09            -10.86   \n",
       "60  TGACTCTCTATAAGGAACTTGAGGATAACC        -5.66      -20.49             -7.01   \n",
       "61  TCTACCTTATGGAAGCTGGAGGTTTCCGTG        -8.81      -24.12            -10.84   \n",
       "62  CTTAAGGACCAACATAAAGGGAGGAGACTC        -9.31      -27.24            -12.79   \n",
       "\n",
       "         iCUB    GC_cds  GC_upstream         viral_id  \n",
       "0   50.525906  0.468927     0.400000  cds-NP_041954.1  \n",
       "1   43.066882  0.455128     0.500000  cds-NP_041955.1  \n",
       "2   55.711182  0.493519     0.366667  cds-NP_041959.1  \n",
       "3   39.155014  0.458333     0.300000  cds-NP_041956.1  \n",
       "6   42.406008  0.481481     0.466667  cds-NP_041958.1  \n",
       "7   53.909493  0.514706     0.400000  cds-NP_041960.1  \n",
       "8   45.974315  0.422481     0.566667  cds-NP_041962.1  \n",
       "9   37.097063  0.480620     0.266667  cds-NP_041961.1  \n",
       "10  55.811993  0.469444     0.266667  cds-NP_041963.1  \n",
       "11  50.382798  0.455128     0.433333  cds-NP_041964.1  \n",
       "12  36.766447  0.444444     0.433333  cds-NP_041965.1  \n",
       "13  48.840195  0.482759     0.366667  cds-NP_041966.1  \n",
       "14  55.478469  0.441624     0.433333  cds-NP_041967.1  \n",
       "15  43.274884  0.469388     0.266667  cds-NP_041968.1  \n",
       "16  45.695741  0.487179     0.400000  cds-NP_041969.1  \n",
       "17  49.136668  0.503577     0.333333  cds-NP_041970.1  \n",
       "18  52.481875  0.490476     0.500000  cds-NP_041971.1  \n",
       "19  54.008683  0.431111     0.466667  cds-NP_041972.1  \n",
       "20  50.637820  0.521930     0.266667  cds-NP_041973.1  \n",
       "21  54.679623  0.456284     0.400000  cds-NP_041974.1  \n",
       "22  56.287505  0.488536     0.433333  cds-NP_041975.1  \n",
       "23  38.086833  0.487805     0.466667  cds-NP_041976.1  \n",
       "24  55.534298  0.490741     0.466667  cds-NP_041977.1  \n",
       "25  54.358884  0.477876     0.433333  cds-NP_041978.1  \n",
       "26  46.132562  0.497653     0.433333  cds-NP_041979.1  \n",
       "27  50.425452  0.485185     0.333333  cds-NP_041980.1  \n",
       "28  46.775101  0.455882     0.366667  cds-NP_041981.1  \n",
       "29  55.203032  0.505910     0.300000  cds-NP_041982.1  \n",
       "30  52.963385  0.425770     0.533333  cds-NP_041983.1  \n",
       "33  50.230438  0.443333     0.266667  cds-NP_041985.1  \n",
       "34  46.284931  0.514286     0.533333  cds-NP_041986.1  \n",
       "35  42.758423  0.452830     0.433333  cds-NP_041987.1  \n",
       "36  52.640637  0.495017     0.400000  cds-NP_041988.1  \n",
       "37  43.669331  0.491228     0.333333  cds-NP_041989.1  \n",
       "38  50.568240  0.443137     0.233333  cds-NP_041990.1  \n",
       "39  50.521668  0.513109     0.433333  cds-NP_041991.1  \n",
       "40  51.947299  0.487562     0.500000  cds-NP_041992.1  \n",
       "41  49.735661  0.513333     0.333333  cds-NP_041993.1  \n",
       "42  52.197060  0.447837     0.400000  cds-NP_041994.1  \n",
       "43  56.023748  0.492241     0.400000  cds-NP_041995.1  \n",
       "44  52.222377  0.515152     0.300000  cds-NP_041996.1  \n",
       "47  48.443366  0.530829     0.233333  cds-NP_041998.1  \n",
       "48  51.781196  0.483926     0.433333  cds-NP_041999.1  \n",
       "49  56.887108  0.470440     0.333333  cds-NP_042000.1  \n",
       "50  51.281910  0.458034     0.433333  cds-NP_042001.1  \n",
       "51  54.614836  0.505922     0.333333  cds-NP_042002.1  \n",
       "52  55.540588  0.500000     0.500000  cds-NP_042003.1  \n",
       "53  56.602989  0.495577     0.433333  cds-NP_042004.1  \n",
       "54  57.179167  0.480746     0.333333  cds-NP_042005.1  \n",
       "55  48.008637  0.441176     0.533333  cds-NP_042006.1  \n",
       "56  46.598805  0.451852     0.400000  cds-NP_042007.1  \n",
       "57  54.753265  0.488426     0.433333  cds-NP_042008.1  \n",
       "58  49.879406  0.500000     0.466667  cds-NP_042009.1  \n",
       "59  54.254624  0.498580     0.400000  cds-NP_042010.1  \n",
       "60  48.602199  0.527132     0.400000  cds-NP_042011.1  \n",
       "61  47.883227  0.482759     0.500000  cds-NP_042012.1  \n",
       "62  51.309820  0.473333     0.466667  cds-NP_042013.1  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "66px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
